{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "source": [
        "1. Assignment: DNA Sequence Analysis. Task: Analyze a given DNA sequence and perform basic sequence\n",
        "manipulation, including finding motifs, calculating GC content, and identifying coding regions. Deliverable: A\n",
        "report summarizing the analysis results and any insights gained from the sequence."
      ],
      "metadata": {
        "id": "QlpYHURctb3r"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "VyqDMo10tZ21"
      },
      "outputs": [],
      "source": [
        "import re\n",
        "sequence = \"\"\n",
        "with open(\"BI_1_sequence.fasta\") as file:\n",
        "  lines = file.readlines()\n",
        "  sequence = ''.join(line.strip() for line in lines[1:]).upper()"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "g_count = sequence.count('G')\n",
        "c_count = sequence.count('C')\n",
        "total_count = len(sequence)\n",
        "gc_percent = ((g_count + c_count) / total_count) * 100\n",
        "print(f\"G Count: {g_count}\")\n",
        "print(f\"C Count: {c_count}\")\n",
        "print(f\"Total Count: {total_count}\")\n",
        "print(f\"GC Percent: {gc_percent:.2f}%\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "3mFm3zZFuNWE",
        "outputId": "4453c24c-1b31-4b8a-a360-b763714416c0"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "G Count: 1130\n",
            "C Count: 1067\n",
            "Total Count: 3598\n",
            "GC Percent: 61.06%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "a_count = sequence.count('A')\n",
        "t_count = sequence.count('T')\n",
        "at_percent = ((a_count + t_count) / total_count) * 100\n",
        "print(f\"A Count: {a_count}\")\n",
        "print(f\"T Count: {t_count}\")\n",
        "print(f\"AT Percent: {at_percent:.2f}%\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "B0G5INTcuNSr",
        "outputId": "65d0780b-72be-465e-aaf0-ff8fdc0faa68"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "A Count: 667\n",
            "T Count: 734\n",
            "AT Percent: 38.94%\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "ratio = (a_count + t_count) / (g_count + c_count)\n",
        "print(f\"AT/GC Ratio: {ratio:.2f}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "eLBDx6v7uNQR",
        "outputId": "c501858a-72da-4737-a2a8-a7ff927814eb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "AT/GC Ratio: 0.64\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "def find_motifs(sequence, motif):\n",
        "    print(f\"\\nSearching for motif: {motif}\")\n",
        "    matches = [match.start() for match in re.finditer(motif, sequence)]\n",
        "    if matches:\n",
        "        print(f\"Motif '{motif}' found at positions: {matches}\")\n",
        "    else:\n",
        "        print(f\"Motif '{motif}' not found\")"
      ],
      "metadata": {
        "id": "oORHaCxsvLH4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "motif = \"TATAA\"\n",
        "find_motifs(sequence, motif)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "lA6DnwxdvK_6",
        "outputId": "333aa2d8-38e3-4f36-92ff-5506320caaa7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Searching for motif: TATAA\n",
            "Motif 'TATAA' found at positions: [1799]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "start_codon = 'ATG'\n",
        "stop_codons = ['TAA', 'TAG', 'TGA']\n",
        "coding_regions = []\n",
        "start_index = sequence.find(start_codon)\n",
        "print(start_index)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WqYtiH6Du0Bx",
        "outputId": "c07560cb-bc63-4bde-d0ac-e9246b872091"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "73\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "while start_index != -1:\n",
        "    for stop_codon in stop_codons:\n",
        "        stop_index = sequence.find(stop_codon, start_index + 3)\n",
        "        if stop_index != -1 and (stop_index - start_index) % 3 == 0:\n",
        "            coding_region = sequence[start_index:stop_index + 3]\n",
        "            coding_regions.append(coding_region)\n",
        "            break\n",
        "    start_index = sequence.find(start_codon, start_index + 1)"
      ],
      "metadata": {
        "id": "nJlq4vriuz-Q"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "if coding_regions:\n",
        "    print(\"Coding Regions Found\")\n",
        "    for i, coding_region in enumerate(coding_regions, start=1):\n",
        "        print(f\"\\nRegion {i}: {coding_region}\\nLength: {len(coding_region)}\")\n",
        "else:\n",
        "    print(\"No coding regions found\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ASgXaQmVuz78",
        "outputId": "9252779f-412a-4987-d457-81274c3a7739"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Coding Regions Found\n",
            "\n",
            "Region 1: ATGAGCTCAGGGGCCTCTAGAAAGAGCTGGGACCCTGGGAACCCCTGGCCTCCAGGTAGTCTCAGGAGAGCTACTCGGGGTCGGGCTTGGGGAGAGGAGGAGCGGGGGTGAGGCAAGCAGCAGGGGACTGGACCTGGGAAGGGCTGGGCAGCAGAGACGACCCGACCCGCTAGAAGGTGGGGTGGGGAGAGCAGCTGGACTGGGATGTAA\n",
            "Length: 210\n",
            "\n",
            "Region 2: ATGTAA\n",
            "Length: 6\n",
            "\n",
            "Region 3: ATGGAACACGGCGCTTAA\n",
            "Length: 18\n",
            "\n",
            "Region 4: ATGTGAAGGGAGAATGAGGAATGCGAGACTGGGACTGAGATGGAACCGGCGGTGGGGAGGGGGTGGGGGGATGGAATTTGAACCCCGGGAGAGGAAGATGGAATTTTCTATGGAGGCCGACCTGGGGATGGGGAGATAAGAGAAGACCAGGAGGGAGTTAAATAG\n",
            "Length: 165\n",
            "\n",
            "Region 5: ATGAGGAATGCGAGACTGGGACTGAGATGGAACCGGCGGTGGGGAGGGGGTGGGGGGATGGAATTTGAACCCCGGGAGAGGAAGATGGAATTTTCTATGGAGGCCGACCTGGGGATGGGGAGATAA\n",
            "Length: 126\n",
            "\n",
            "Region 6: ATGCGAGACTGGGACTGA\n",
            "Length: 18\n",
            "\n",
            "Region 7: ATGGAACCGGCGGTGGGGAGGGGGTGGGGGGATGGAATTTGAACCCCGGGAGAGGAAGATGGAATTTTCTATGGAGGCCGACCTGGGGATGGGGAGATAAGAGAAGACCAGGAGGGAGTTAAATAG\n",
            "Length: 126\n",
            "\n",
            "Region 8: ATGGAATTTGAACCCCGGGAGAGGAAGATGGAATTTTCTATGGAGGCCGACCTGGGGATGGGGAGATAA\n",
            "Length: 69\n",
            "\n",
            "Region 9: ATGGAATTTTCTATGGAGGCCGACCTGGGGATGGGGAGATAA\n",
            "Length: 42\n",
            "\n",
            "Region 10: ATGGAGGCCGACCTGGGGATGGGGAGATAA\n",
            "Length: 30\n",
            "\n",
            "Region 11: ATGGGGAGATAA\n",
            "Length: 12\n",
            "\n",
            "Region 12: ATGTGCTGGGATTAG\n",
            "Length: 15\n",
            "\n",
            "Region 13: ATGCAACAAGGCTTGGAAGGCTAA\n",
            "Length: 24\n",
            "\n",
            "Region 14: ATGAAGGTTCTGTGGGCTGCGTTGCTGGTCACATTCCTGGCAGGTATGGGGGCGGGGCTTGCTCGGTTCCCCCCGCTCCTCCCCCTCTCATCCTCACCTCAACCTCCTGGCCCCATTCAGGCAGACCCTGGGCCCCCTCTTCTGAGGCTTCTGTGCTGCTTCCTGGCTCTGAACAGCGATTTGACGCTCTCTGGGCCTCGGTTTCCCCCATCCTTGAGATAGGAGTTAGAAGTTGTTTTGTTGTTGTTGTTTGTTGTTGTTGTTTTGTTTTTTTGAGATGAAGTCTCGCTCTGTCGCCCAGGCTGGAGTGCAGTGGCGGGATCTCGGCTCACTGCAAGCTCCGCCTCCCAGGTCCACGCCATTCTCCTGCCTCAGCCTCCCAAGTAGCTGGGACTACAGGCACATGCCACCACACCCGACTAA\n",
            "Length: 423\n",
            "\n",
            "Region 15: ATGGGGGCGGGGCTTGCTCGGTTCCCCCCGCTCCTCCCCCTCTCATCCTCACCTCAACCTCCTGGCCCCATTCAGGCAGACCCTGGGCCCCCTCTTCTGAGGCTTCTGTGCTGCTTCCTGGCTCTGAACAGCGATTTGACGCTCTCTGGGCCTCGGTTTCCCCCATCCTTGAGATAGGAGTTAGAAGTTGTTTTGTTGTTGTTGTTTGTTGTTGTTGTTTTGTTTTTTTGAGATGAAGTCTCGCTCTGTCGCCCAGGCTGGAGTGCAGTGGCGGGATCTCGGCTCACTGCAAGCTCCGCCTCCCAGGTCCACGCCATTCTCCTGCCTCAGCCTCCCAAGTAGCTGGGACTACAGGCACATGCCACCACACCCGACTAA\n",
            "Length: 378\n",
            "\n",
            "Region 16: ATGTTGGCCAGGCTGGTCTGGAACTCCTGA\n",
            "Length: 30\n",
            "\n",
            "Region 17: ATGCTTTTCCGCTGGGCGCGGTGGCTCACCCCTGTAATCCCAGCACTTTGGGAGGCCAAGGTGGGAGGATCACTTGAGCCCAGGAGTTCAACACCAGCCTGGGCAACATAG\n",
            "Length: 111\n",
            "\n",
            "Region 18: ATGGTGCCACACACCTGTGCTCTCAGCTACTCAGGAGGCTGA\n",
            "Length: 42\n",
            "\n",
            "Region 19: ATGTTCAGGCCGCTGCACTCCAGCCTGGGTGACAGAGCAAGACCCTGTTTATAA\n",
            "Length: 54\n",
            "\n",
            "Region 20: ATGCTTTCCAAGTGA\n",
            "Length: 15\n",
            "\n",
            "Region 21: ATGCCTGGACGGGGTCAGAAGGACCCTGACCCACCTTGAACTTGTTCCACACAGGATGCCAGGCCAAGGTGGAGCAAGCGGTGGAGACAGAGCCGGAGCCCGAGCTGCGCCAGCAGACCGAGTGGCAGAGCGGCCAGCGCTGGGAACTGGCACTGGGTCGCTTTTGGGATTACCTGCGCTGGGTGCAGACACTGTCTGAGCAGGTGCAGGAGGAGCTGCTCAGCTCCCAGGTCACCCAGGAACTGAGGTGAGTGTCCCCATCCTGGCCCTTGACCCTCCTGGTGGGCGGCTATACCTCCCCAGGTCCAGGTTTCATTCTGCCCCTGTCGCTAA\n",
            "Length: 333\n",
            "\n",
            "Region 22: ATGCCAGGCCAAGGTGGAGCAAGCGGTGGAGACAGAGCCGGAGCCCGAGCTGCGCCAGCAGACCGAGTGGCAGAGCGGCCAGCGCTGGGAACTGGCACTGGGTCGCTTTTGGGATTACCTGCGCTGGGTGCAGACACTGTCTGA\n",
            "Length: 144\n",
            "\n",
            "Region 23: ATGAGCCACCTTGCCCGGCCTCCTAGCTCCTTCTTCGTCTCTGCCTCTGCCCTCTGCATCTGCTCTCTGCATCTGTCTCTGTCTCCTTCTCTCGGCCTCTGCCCCGTTCCTTCTCTCCCTCTTGGGTCTCTCTGGCTCATCCCCATCTCGCCCGCCCCATCCCAGCCCTTCTCCCCGCCTCCCACTGTGCGACACCCTCCCGCCCTCTCGGCCGCAGGGCGCTGA\n",
            "Length: 225\n",
            "\n",
            "Region 24: ATGGACGAGACCATGAAGGAGTTGAAGGCCTACAAATCGGAACTGGAGGAACAACTGACCCCGGTGGCGGAGGAGACGCGGGCACGGCTGTCCAAGGAGCTGCAGGCGGCGCAGGCCCGGCTGGGCGCGGACATGGAGGACGTGTGCGGCCGCCTGGTGCAGTACCGCGGCGAGGTGCAGGCCATGCTCGGCCAGAGCACCGAGGAGCTGCGGGTGCGCCTCGCCTCCCACCTGCGCAAGCTGCGTAAGCGGCTCCTCCGCGATGCCGATGACCTGCAGAAGCGCCTGGCAGTGTACCAGGCCGGGGCCCGCGAGGGCGCCGAGCGCGGCCTCAGCGCCATCCGCGAGCGCCTGGGGCCCCTGGTGGAACAGGGCCGCGTGCGGGCCGCCACTGTGGGCTCCCTGGCCGGCCAGCCGCTACAGGAGCGGGCCCAGGCCTGGGGCGAGCGGCTGCGCGCGCGGATGGAGGAGATGGGCAGCCGGACCCGCGACCGCCTGGACGAGGTGAAGGAGCAGGTGGCGGAGGTGCGCGCCAAGCTGGAGGAGCAGGCCCAGCAGATACGCCTGCAGGCCGAGGCCTTCCAGGCCCGCCTCAAGAGCTGGTTCGAGCCCCTGGTGGAAGACATGCAGCGCCAGTGGGCCGGGCTGGTGGAGAAGGTGCAGGCTGCCGTGGGCACCAGCGCCGCCCCTGTGCCCAGCGACAATCACTGAACGCCGAAGCCTGCAGCCATGCGACCCCACGCCACCCCGTGCCTCCTGCCTCCGCGCAGCCTGCAGCGGGAGACCCTGTCCCCGCCCCAGCCGTCCTCCTGGGGTGGACCCTAG\n",
            "Length: 825\n",
            "\n",
            "Region 25: ATGAAGGAGTTGAAGGCCTACAAATCGGAACTGGAGGAACAACTGACCCCGGTGGCGGAGGAGACGCGGGCACGGCTGTCCAAGGAGCTGCAGGCGGCGCAGGCCCGGCTGGGCGCGGACATGGAGGACGTGTGCGGCCGCCTGGTGCAGTACCGCGGCGAGGTGCAGGCCATGCTCGGCCAGAGCACCGAGGAGCTGCGGGTGCGCCTCGCCTCCCACCTGCGCAAGCTGCGTAAGCGGCTCCTCCGCGATGCCGATGACCTGCAGAAGCGCCTGGCAGTGTACCAGGCCGGGGCCCGCGAGGGCGCCGAGCGCGGCCTCAGCGCCATCCGCGAGCGCCTGGGGCCCCTGGTGGAACAGGGCCGCGTGCGGGCCGCCACTGTGGGCTCCCTGGCCGGCCAGCCGCTACAGGAGCGGGCCCAGGCCTGGGGCGAGCGGCTGCGCGCGCGGATGGAGGAGATGGGCAGCCGGACCCGCGACCGCCTGGACGAGGTGAAGGAGCAGGTGGCGGAGGTGCGCGCCAAGCTGGAGGAGCAGGCCCAGCAGATACGCCTGCAGGCCGAGGCCTTCCAGGCCCGCCTCAAGAGCTGGTTCGAGCCCCTGGTGGAAGACATGCAGCGCCAGTGGGCCGGGCTGGTGGAGAAGGTGCAGGCTGCCGTGGGCACCAGCGCCGCCCCTGTGCCCAGCGACAATCACTGAACGCCGAAGCCTGCAGCCATGCGACCCCACGCCACCCCGTGCCTCCTGCCTCCGCGCAGCCTGCAGCGGGAGACCCTGTCCCCGCCCCAGCCGTCCTCCTGGGGTGGACCCTAG\n",
            "Length: 813\n",
            "\n",
            "Region 26: ATGGAGGACGTGTGCGGCCGCCTGGTGCAGTACCGCGGCGAGGTGCAGGCCATGCTCGGCCAGAGCACCGAGGAGCTGCGGGTGCGCCTCGCCTCCCACCTGCGCAAGCTGCGTAAGCGGCTCCTCCGCGATGCCGATGACCTGCAGAAGCGCCTGGCAGTGTACCAGGCCGGGGCCCGCGAGGGCGCCGAGCGCGGCCTCAGCGCCATCCGCGAGCGCCTGGGGCCCCTGGTGGAACAGGGCCGCGTGCGGGCCGCCACTGTGGGCTCCCTGGCCGGCCAGCCGCTACAGGAGCGGGCCCAGGCCTGGGGCGAGCGGCTGCGCGCGCGGATGGAGGAGATGGGCAGCCGGACCCGCGACCGCCTGGACGAGGTGAAGGAGCAGGTGGCGGAGGTGCGCGCCAAGCTGGAGGAGCAGGCCCAGCAGATACGCCTGCAGGCCGAGGCCTTCCAGGCCCGCCTCAAGAGCTGGTTCGAGCCCCTGGTGGAAGACATGCAGCGCCAGTGGGCCGGGCTGGTGGAGAAGGTGCAGGCTGCCGTGGGCACCAGCGCCGCCCCTGTGCCCAGCGACAATCACTGAACGCCGAAGCCTGCAGCCATGCGACCCCACGCCACCCCGTGCCTCCTGCCTCCGCGCAGCCTGCAGCGGGAGACCCTGTCCCCGCCCCAGCCGTCCTCCTGGGGTGGACCCTAG\n",
            "Length: 693\n",
            "\n",
            "Region 27: ATGCTCGGCCAGAGCACCGAGGAGCTGCGGGTGCGCCTCGCCTCCCACCTGCGCAAGCTGCGTAAGCGGCTCCTCCGCGATGCCGATGACCTGCAGAAGCGCCTGGCAGTGTACCAGGCCGGGGCCCGCGAGGGCGCCGAGCGCGGCCTCAGCGCCATCCGCGAGCGCCTGGGGCCCCTGGTGGAACAGGGCCGCGTGCGGGCCGCCACTGTGGGCTCCCTGGCCGGCCAGCCGCTACAGGAGCGGGCCCAGGCCTGGGGCGAGCGGCTGCGCGCGCGGATGGAGGAGATGGGCAGCCGGACCCGCGACCGCCTGGACGAGGTGAAGGAGCAGGTGGCGGAGGTGCGCGCCAAGCTGGAGGAGCAGGCCCAGCAGATACGCCTGCAGGCCGAGGCCTTCCAGGCCCGCCTCAAGAGCTGGTTCGAGCCCCTGGTGGAAGACATGCAGCGCCAGTGGGCCGGGCTGGTGGAGAAGGTGCAGGCTGCCGTGGGCACCAGCGCCGCCCCTGTGCCCAGCGACAATCACTGAACGCCGAAGCCTGCAGCCATGCGACCCCACGCCACCCCGTGCCTCCTGCCTCCGCGCAGCCTGCAGCGGGAGACCCTGTCCCCGCCCCAGCCGTCCTCCTGGGGTGGACCCTAG\n",
            "Length: 642\n",
            "\n",
            "Region 28: ATGACCTGCAGAAGCGCCTGGCAGTGTACCAGGCCGGGGCCCGCGAGGGCGCCGAGCGCGGCCTCAGCGCCATCCGCGAGCGCCTGGGGCCCCTGGTGGAACAGGGCCGCGTGCGGGCCGCCACTGTGGGCTCCCTGGCCGGCCAGCCGCTACAGGAGCGGGCCCAGGCCTGGGGCGAGCGGCTGCGCGCGCGGATGGAGGAGATGGGCAGCCGGACCCGCGACCGCCTGGACGAGGTGA\n",
            "Length: 240\n",
            "\n",
            "Region 29: ATGGAGGAGATGGGCAGCCGGACCCGCGACCGCCTGGACGAGGTGAAGGAGCAGGTGGCGGAGGTGCGCGCCAAGCTGGAGGAGCAGGCCCAGCAGATACGCCTGCAGGCCGAGGCCTTCCAGGCCCGCCTCAAGAGCTGGTTCGAGCCCCTGGTGGAAGACATGCAGCGCCAGTGGGCCGGGCTGGTGGAGAAGGTGCAGGCTGCCGTGGGCACCAGCGCCGCCCCTGTGCCCAGCGACAATCACTGAACGCCGAAGCCTGCAGCCATGCGACCCCACGCCACCCCGTGCCTCCTGCCTCCGCGCAGCCTGCAGCGGGAGACCCTGTCCCCGCCCCAGCCGTCCTCCTGGGGTGGACCCTAG\n",
            "Length: 363\n",
            "\n",
            "Region 30: ATGGGCAGCCGGACCCGCGACCGCCTGGACGAGGTGAAGGAGCAGGTGGCGGAGGTGCGCGCCAAGCTGGAGGAGCAGGCCCAGCAGATACGCCTGCAGGCCGAGGCCTTCCAGGCCCGCCTCAAGAGCTGGTTCGAGCCCCTGGTGGAAGACATGCAGCGCCAGTGGGCCGGGCTGGTGGAGAAGGTGCAGGCTGCCGTGGGCACCAGCGCCGCCCCTGTGCCCAGCGACAATCACTGAACGCCGAAGCCTGCAGCCATGCGACCCCACGCCACCCCGTGCCTCCTGCCTCCGCGCAGCCTGCAGCGGGAGACCCTGTCCCCGCCCCAGCCGTCCTCCTGGGGTGGACCCTAG\n",
            "Length: 354\n",
            "\n",
            "Region 31: ATGCAGCGCCAGTGGGCCGGGCTGGTGGAGAAGGTGCAGGCTGCCGTGGGCACCAGCGCCGCCCCTGTGCCCAGCGACAATCACTGAACGCCGAAGCCTGCAGCCATGCGACCCCACGCCACCCCGTGCCTCCTGCCTCCGCGCAGCCTGCAGCGGGAGACCCTGTCCCCGCCCCAGCCGTCCTCCTGGGGTGGACCCTAG\n",
            "Length: 201\n",
            "\n",
            "Region 32: ATGCGACCCCACGCCACCCCGTGCCTCCTGCCTCCGCGCAGCCTGCAGCGGGAGACCCTGTCCCCGCCCCAGCCGTCCTCCTGGGGTGGACCCTAG\n",
            "Length: 96\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "2. Assignment: RNA-Seq Data Analysis. Task: Analyze a provided RNA-Seq dataset and perform differential\n",
        "gene expression analysis. Deliverable: A detailed report presenting the differentially expressed genes, their\n",
        "functional annotations, and any potential biological interpretations"
      ],
      "metadata": {
        "id": "OofFpSH_uM9G"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#R Studio"
      ],
      "metadata": {
        "id": "37TFOdDmuenx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "3. Assignment: Protein Structure Prediction. Task: Predict the 3D structure of a given protein sequence using\n",
        "homology modeling or threading techniques. Deliverable: A report presenting the predicted protein structure,\n",
        "along with an analysis of its potential functions and interactions."
      ],
      "metadata": {
        "id": "t1BM4gd0uMvc"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "#Tools"
      ],
      "metadata": {
        "id": "SdSqhb5Yud4M"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "5. Assignment: Machine Learning for Genomic Data. Task: Apply machine learning algorithms, such as random\n",
        "forests or support vector machines, to classify genomic data based on specific features or markers. Deliverable:\n",
        "A comprehensive analysis report presenting the classification results, model performance evaluation, and\n",
        "insights into the predictive features."
      ],
      "metadata": {
        "id": "RcQFYwL8uWZw"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Step 1: Import necessary libraries\n",
        "import pandas as pd\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.ensemble import RandomForestClassifier\n",
        "from sklearn.metrics import accuracy_score, confusion_matrix, classification_report, precision_score, recall_score\n",
        "from sklearn.preprocessing import LabelEncoder\n",
        "import matplotlib.pyplot as plt"
      ],
      "metadata": {
        "id": "sDNhxesaudaO"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Step 2: Load the dataset\n",
        "data = pd.read_csv('METABRIC_RNA_Mutation.csv')"
      ],
      "metadata": {
        "id": "lABPIE_hudWo"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Step 3: Explore the dataset\n",
        "print(data.head())"
      ],
      "metadata": {
        "id": "ljH67_smucza",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "230615d6-2b9a-4334-b1d3-0e951476ca31"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "   patient_id  age_at_diagnosis type_of_breast_surgery    cancer_type  \\\n",
            "0           0             75.65             MASTECTOMY  Breast Cancer   \n",
            "1           2             43.19      BREAST CONSERVING  Breast Cancer   \n",
            "2           5             48.87             MASTECTOMY  Breast Cancer   \n",
            "3           6             47.68             MASTECTOMY  Breast Cancer   \n",
            "4           8             76.97             MASTECTOMY  Breast Cancer   \n",
            "\n",
            "                        cancer_type_detailed cellularity  chemotherapy  \\\n",
            "0           Breast Invasive Ductal Carcinoma         NaN             0   \n",
            "1           Breast Invasive Ductal Carcinoma        High             0   \n",
            "2           Breast Invasive Ductal Carcinoma        High             1   \n",
            "3  Breast Mixed Ductal and Lobular Carcinoma    Moderate             1   \n",
            "4  Breast Mixed Ductal and Lobular Carcinoma        High             1   \n",
            "\n",
            "  pam50_+_claudin-low_subtype  cohort er_status_measured_by_ihc  ... mtap_mut  \\\n",
            "0                 claudin-low     1.0                   Positve  ...        0   \n",
            "1                        LumA     1.0                   Positve  ...        0   \n",
            "2                        LumB     1.0                   Positve  ...        0   \n",
            "3                        LumB     1.0                   Positve  ...        0   \n",
            "4                        LumB     1.0                   Positve  ...        0   \n",
            "\n",
            "   ppp2cb_mut smarcd1_mut nras_mut ndfip1_mut  hras_mut prps2_mut smarcb1_mut  \\\n",
            "0           0           0        0          0       0.0         0         0.0   \n",
            "1           0           0        0          0       0.0         0         0.0   \n",
            "2           0           0        0          0       0.0         0         0.0   \n",
            "3           0           0        0          0       0.0         0         0.0   \n",
            "4           0           0        0          0       0.0         0         0.0   \n",
            "\n",
            "  stmn2_mut  siah1_mut  \n",
            "0         0        0.0  \n",
            "1         0        0.0  \n",
            "2         0        0.0  \n",
            "3         0        0.0  \n",
            "4         0        0.0  \n",
            "\n",
            "[5 rows x 693 columns]\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data.info(verbose=True)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ysLrWvaPpbzi",
        "outputId": "02cba127-8e03-4d25-a2a1-1f9581501b26"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "<class 'pandas.core.frame.DataFrame'>\n",
            "RangeIndex: 715 entries, 0 to 714\n",
            "Data columns (total 693 columns):\n",
            " #    Column                          Dtype  \n",
            "---   ------                          -----  \n",
            " 0    patient_id                      int64  \n",
            " 1    age_at_diagnosis                float64\n",
            " 2    type_of_breast_surgery          object \n",
            " 3    cancer_type                     object \n",
            " 4    cancer_type_detailed            object \n",
            " 5    cellularity                     object \n",
            " 6    chemotherapy                    int64  \n",
            " 7    pam50_+_claudin-low_subtype     object \n",
            " 8    cohort                          float64\n",
            " 9    er_status_measured_by_ihc       object \n",
            " 10   er_status                       object \n",
            " 11   neoplasm_histologic_grade       float64\n",
            " 12   her2_status_measured_by_snp6    object \n",
            " 13   her2_status                     object \n",
            " 14   tumor_other_histologic_subtype  object \n",
            " 15   hormone_therapy                 int64  \n",
            " 16   inferred_menopausal_state       object \n",
            " 17   integrative_cluster             object \n",
            " 18   primary_tumor_laterality        object \n",
            " 19   lymph_nodes_examined_positive   float64\n",
            " 20   mutation_count                  float64\n",
            " 21   nottingham_prognostic_index     float64\n",
            " 22   oncotree_code                   object \n",
            " 23   overall_survival_months         float64\n",
            " 24   overall_survival                int64  \n",
            " 25   pr_status                       object \n",
            " 26   radio_therapy                   int64  \n",
            " 27   3-gene_classifier_subtype       object \n",
            " 28   tumor_size                      float64\n",
            " 29   tumor_stage                     float64\n",
            " 30   death_from_cancer               object \n",
            " 31   brca1                           float64\n",
            " 32   brca2                           float64\n",
            " 33   palb2                           float64\n",
            " 34   pten                            float64\n",
            " 35   tp53                            float64\n",
            " 36   atm                             float64\n",
            " 37   cdh1                            float64\n",
            " 38   chek2                           float64\n",
            " 39   nbn                             float64\n",
            " 40   nf1                             float64\n",
            " 41   stk11                           float64\n",
            " 42   bard1                           float64\n",
            " 43   mlh1                            float64\n",
            " 44   msh2                            float64\n",
            " 45   msh6                            float64\n",
            " 46   pms2                            float64\n",
            " 47   epcam                           float64\n",
            " 48   rad51c                          float64\n",
            " 49   rad51d                          float64\n",
            " 50   rad50                           float64\n",
            " 51   rb1                             float64\n",
            " 52   rbl1                            float64\n",
            " 53   rbl2                            float64\n",
            " 54   ccna1                           float64\n",
            " 55   ccnb1                           float64\n",
            " 56   cdk1                            float64\n",
            " 57   ccne1                           float64\n",
            " 58   cdk2                            float64\n",
            " 59   cdc25a                          float64\n",
            " 60   ccnd1                           float64\n",
            " 61   cdk4                            float64\n",
            " 62   cdk6                            float64\n",
            " 63   ccnd2                           float64\n",
            " 64   cdkn2a                          float64\n",
            " 65   cdkn2b                          float64\n",
            " 66   myc                             float64\n",
            " 67   cdkn1a                          float64\n",
            " 68   cdkn1b                          float64\n",
            " 69   e2f1                            float64\n",
            " 70   e2f2                            float64\n",
            " 71   e2f3                            float64\n",
            " 72   e2f4                            float64\n",
            " 73   e2f5                            float64\n",
            " 74   e2f6                            float64\n",
            " 75   e2f7                            float64\n",
            " 76   e2f8                            float64\n",
            " 77   src                             float64\n",
            " 78   jak1                            float64\n",
            " 79   jak2                            float64\n",
            " 80   stat1                           float64\n",
            " 81   stat2                           float64\n",
            " 82   stat3                           float64\n",
            " 83   stat5a                          float64\n",
            " 84   stat5b                          float64\n",
            " 85   mdm2                            float64\n",
            " 86   tp53bp1                         float64\n",
            " 87   adam10                          float64\n",
            " 88   adam17                          float64\n",
            " 89   aph1a                           float64\n",
            " 90   aph1b                           float64\n",
            " 91   arrdc1                          float64\n",
            " 92   cir1                            float64\n",
            " 93   ctbp1                           float64\n",
            " 94   ctbp2                           float64\n",
            " 95   cul1                            float64\n",
            " 96   dll1                            float64\n",
            " 97   dll3                            float64\n",
            " 98   dll4                            float64\n",
            " 99   dtx1                            float64\n",
            " 100  dtx2                            float64\n",
            " 101  dtx3                            float64\n",
            " 102  dtx4                            float64\n",
            " 103  ep300                           float64\n",
            " 104  fbxw7                           float64\n",
            " 105  hdac1                           float64\n",
            " 106  hdac2                           float64\n",
            " 107  hes1                            float64\n",
            " 108  hes5                            float64\n",
            " 109  heyl                            float64\n",
            " 110  itch                            float64\n",
            " 111  jag1                            float64\n",
            " 112  jag2                            float64\n",
            " 113  kdm5a                           float64\n",
            " 114  lfng                            float64\n",
            " 115  maml1                           float64\n",
            " 116  maml2                           float64\n",
            " 117  maml3                           float64\n",
            " 118  ncor2                           float64\n",
            " 119  ncstn                           float64\n",
            " 120  notch1                          float64\n",
            " 121  notch2                          float64\n",
            " 122  notch3                          float64\n",
            " 123  nrarp                           float64\n",
            " 124  numb                            float64\n",
            " 125  numbl                           float64\n",
            " 126  psen1                           float64\n",
            " 127  psen2                           float64\n",
            " 128  psenen                          float64\n",
            " 129  rbpj                            float64\n",
            " 130  rbpjl                           float64\n",
            " 131  rfng                            float64\n",
            " 132  snw1                            float64\n",
            " 133  spen                            float64\n",
            " 134  hes2                            float64\n",
            " 135  hes4                            float64\n",
            " 136  hes7                            float64\n",
            " 137  hey1                            float64\n",
            " 138  hey2                            float64\n",
            " 139  acvr1                           float64\n",
            " 140  acvr1b                          float64\n",
            " 141  acvr1c                          float64\n",
            " 142  acvr2a                          float64\n",
            " 143  acvr2b                          float64\n",
            " 144  acvrl1                          float64\n",
            " 145  akt1                            float64\n",
            " 146  akt1s1                          float64\n",
            " 147  akt2                            float64\n",
            " 148  apaf1                           float64\n",
            " 149  arl11                           float64\n",
            " 150  atr                             float64\n",
            " 151  aurka                           float64\n",
            " 152  bad                             float64\n",
            " 153  bcl2                            float64\n",
            " 154  bcl2l1                          float64\n",
            " 155  bmp10                           float64\n",
            " 156  bmp15                           float64\n",
            " 157  bmp2                            float64\n",
            " 158  bmp3                            float64\n",
            " 159  bmp4                            float64\n",
            " 160  bmp5                            float64\n",
            " 161  bmp6                            float64\n",
            " 162  bmp7                            float64\n",
            " 163  bmpr1a                          float64\n",
            " 164  bmpr1b                          float64\n",
            " 165  bmpr2                           float64\n",
            " 166  braf                            float64\n",
            " 167  casp10                          float64\n",
            " 168  casp3                           float64\n",
            " 169  casp6                           float64\n",
            " 170  casp7                           float64\n",
            " 171  casp8                           float64\n",
            " 172  casp9                           float64\n",
            " 173  chek1                           float64\n",
            " 174  csf1                            float64\n",
            " 175  csf1r                           float64\n",
            " 176  cxcl8                           float64\n",
            " 177  cxcr1                           float64\n",
            " 178  cxcr2                           float64\n",
            " 179  dab2                            float64\n",
            " 180  diras3                          float64\n",
            " 181  dlec1                           float64\n",
            " 182  dph1                            float64\n",
            " 183  egfr                            float64\n",
            " 184  eif4e                           float64\n",
            " 185  eif4ebp1                        float64\n",
            " 186  eif5a2                          float64\n",
            " 187  erbb2                           float64\n",
            " 188  erbb3                           float64\n",
            " 189  erbb4                           float64\n",
            " 190  fas                             float64\n",
            " 191  fgf1                            float64\n",
            " 192  fgfr1                           float64\n",
            " 193  folr1                           float64\n",
            " 194  folr2                           float64\n",
            " 195  folr3                           float64\n",
            " 196  foxo1                           float64\n",
            " 197  foxo3                           float64\n",
            " 198  gdf11                           float64\n",
            " 199  gdf2                            float64\n",
            " 200  gsk3b                           float64\n",
            " 201  hif1a                           float64\n",
            " 202  hla-g                           float64\n",
            " 203  hras                            float64\n",
            " 204  igf1                            float64\n",
            " 205  igf1r                           float64\n",
            " 206  inha                            float64\n",
            " 207  inhba                           float64\n",
            " 208  inhbc                           float64\n",
            " 209  itgav                           float64\n",
            " 210  itgb3                           float64\n",
            " 211  izumo1r                         float64\n",
            " 212  kdr                             float64\n",
            " 213  kit                             float64\n",
            " 214  kras                            float64\n",
            " 215  map2k1                          float64\n",
            " 216  map2k2                          float64\n",
            " 217  map2k3                          float64\n",
            " 218  map2k4                          float64\n",
            " 219  map2k5                          float64\n",
            " 220  map3k1                          float64\n",
            " 221  map3k3                          float64\n",
            " 222  map3k4                          float64\n",
            " 223  map3k5                          float64\n",
            " 224  mapk1                           float64\n",
            " 225  mapk12                          float64\n",
            " 226  mapk14                          float64\n",
            " 227  mapk3                           float64\n",
            " 228  mapk4                           float64\n",
            " 229  mapk6                           float64\n",
            " 230  mapk7                           float64\n",
            " 231  mapk8                           float64\n",
            " 232  mapk9                           float64\n",
            " 233  mdc1                            float64\n",
            " 234  mlst8                           float64\n",
            " 235  mmp1                            float64\n",
            " 236  mmp10                           float64\n",
            " 237  mmp11                           float64\n",
            " 238  mmp12                           float64\n",
            " 239  mmp13                           float64\n",
            " 240  mmp14                           float64\n",
            " 241  mmp15                           float64\n",
            " 242  mmp16                           float64\n",
            " 243  mmp17                           float64\n",
            " 244  mmp19                           float64\n",
            " 245  mmp2                            float64\n",
            " 246  mmp21                           float64\n",
            " 247  mmp23b                          float64\n",
            " 248  mmp24                           float64\n",
            " 249  mmp25                           float64\n",
            " 250  mmp26                           float64\n",
            " 251  mmp27                           float64\n",
            " 252  mmp28                           float64\n",
            " 253  mmp3                            float64\n",
            " 254  mmp7                            float64\n",
            " 255  mmp9                            float64\n",
            " 256  mtor                            float64\n",
            " 257  nfkb1                           float64\n",
            " 258  nfkb2                           float64\n",
            " 259  opcml                           float64\n",
            " 260  pdgfa                           float64\n",
            " 261  pdgfb                           float64\n",
            " 262  pdgfra                          float64\n",
            " 263  pdgfrb                          float64\n",
            " 264  pdpk1                           float64\n",
            " 265  peg3                            float64\n",
            " 266  pik3ca                          float64\n",
            " 267  pik3r1                          float64\n",
            " 268  pik3r2                          float64\n",
            " 269  plagl1                          float64\n",
            " 270  ptk2                            float64\n",
            " 271  rab25                           float64\n",
            " 272  rad51                           float64\n",
            " 273  raf1                            float64\n",
            " 274  rassf1                          float64\n",
            " 275  rheb                            float64\n",
            " 276  rictor                          float64\n",
            " 277  rps6                            float64\n",
            " 278  rps6ka1                         float64\n",
            " 279  rps6ka2                         float64\n",
            " 280  rps6kb1                         float64\n",
            " 281  rps6kb2                         float64\n",
            " 282  rptor                           float64\n",
            " 283  slc19a1                         float64\n",
            " 284  smad1                           float64\n",
            " 285  smad2                           float64\n",
            " 286  smad3                           float64\n",
            " 287  smad4                           float64\n",
            " 288  smad5                           float64\n",
            " 289  smad6                           float64\n",
            " 290  smad7                           float64\n",
            " 291  smad9                           float64\n",
            " 292  sptbn1                          float64\n",
            " 293  terc                            float64\n",
            " 294  tert                            float64\n",
            " 295  tgfb1                           float64\n",
            " 296  tgfb2                           float64\n",
            " 297  tgfb3                           float64\n",
            " 298  tgfbr1                          float64\n",
            " 299  tgfbr2                          float64\n",
            " 300  tgfbr3                          float64\n",
            " 301  tsc1                            float64\n",
            " 302  tsc2                            float64\n",
            " 303  vegfa                           float64\n",
            " 304  vegfb                           float64\n",
            " 305  wfdc2                           float64\n",
            " 306  wwox                            float64\n",
            " 307  zfyve9                          float64\n",
            " 308  arid1a                          float64\n",
            " 309  arid1b                          float64\n",
            " 310  cbfb                            float64\n",
            " 311  gata3                           float64\n",
            " 312  kmt2c                           float64\n",
            " 313  kmt2d                           float64\n",
            " 314  myh9                            float64\n",
            " 315  ncor1                           float64\n",
            " 316  pde4dip                         float64\n",
            " 317  ptprd                           float64\n",
            " 318  ros1                            float64\n",
            " 319  runx1                           float64\n",
            " 320  tbx3                            float64\n",
            " 321  abcb1                           float64\n",
            " 322  abcb11                          float64\n",
            " 323  abcc1                           float64\n",
            " 324  abcc10                          float64\n",
            " 325  bbc3                            float64\n",
            " 326  bmf                             float64\n",
            " 327  cyp2c8                          float64\n",
            " 328  cyp3a4                          float64\n",
            " 329  fgf2                            float64\n",
            " 330  fn1                             float64\n",
            " 331  map2                            float64\n",
            " 332  map4                            float64\n",
            " 333  mapt                            float64\n",
            " 334  nr1i2                           float64\n",
            " 335  slco1b3                         float64\n",
            " 336  tubb1                           float64\n",
            " 337  tubb4a                          float64\n",
            " 338  tubb4b                          float64\n",
            " 339  twist1                          float64\n",
            " 340  adgra2                          float64\n",
            " 341  afdn                            float64\n",
            " 342  aff2                            float64\n",
            " 343  agmo                            float64\n",
            " 344  agtr2                           float64\n",
            " 345  ahnak                           float64\n",
            " 346  ahnak2                          float64\n",
            " 347  akap9                           float64\n",
            " 348  alk                             float64\n",
            " 349  apc                             float64\n",
            " 350  arid2                           float64\n",
            " 351  arid5b                          float64\n",
            " 352  asxl1                           float64\n",
            " 353  asxl2                           float64\n",
            " 354  bap1                            float64\n",
            " 355  bcas3                           float64\n",
            " 356  birc6                           float64\n",
            " 357  cacna2d3                        float64\n",
            " 358  ccnd3                           float64\n",
            " 359  chd1                            float64\n",
            " 360  clk3                            float64\n",
            " 361  clrn2                           float64\n",
            " 362  col12a1                         float64\n",
            " 363  col22a1                         float64\n",
            " 364  col6a3                          float64\n",
            " 365  ctcf                            float64\n",
            " 366  ctnna1                          float64\n",
            " 367  ctnna3                          float64\n",
            " 368  dnah11                          float64\n",
            " 369  dnah2                           float64\n",
            " 370  dnah5                           float64\n",
            " 371  dtwd2                           float64\n",
            " 372  fam20c                          float64\n",
            " 373  fanca                           float64\n",
            " 374  fancd2                          float64\n",
            " 375  flt3                            float64\n",
            " 376  foxp1                           float64\n",
            " 377  frmd3                           float64\n",
            " 378  gh1                             float64\n",
            " 379  gldc                            float64\n",
            " 380  gpr32                           float64\n",
            " 381  gps2                            float64\n",
            " 382  hdac9                           float64\n",
            " 383  herc2                           float64\n",
            " 384  hist1h2bc                       float64\n",
            " 385  kdm3a                           float64\n",
            " 386  kdm6a                           float64\n",
            " 387  klrg1                           float64\n",
            " 388  l1cam                           float64\n",
            " 389  lama2                           float64\n",
            " 390  lamb3                           float64\n",
            " 391  large1                          float64\n",
            " 392  ldlrap1                         float64\n",
            " 393  lifr                            float64\n",
            " 394  lipi                            float64\n",
            " 395  magea8                          float64\n",
            " 396  map3k10                         float64\n",
            " 397  map3k13                         float64\n",
            " 398  men1                            float64\n",
            " 399  mtap                            float64\n",
            " 400  muc16                           float64\n",
            " 401  myo1a                           float64\n",
            " 402  myo3a                           float64\n",
            " 403  ncoa3                           float64\n",
            " 404  nek1                            float64\n",
            " 405  nf2                             float64\n",
            " 406  npnt                            float64\n",
            " 407  nr2f1                           float64\n",
            " 408  nr3c1                           float64\n",
            " 409  nras                            float64\n",
            " 410  nrg3                            float64\n",
            " 411  nt5e                            float64\n",
            " 412  or6a2                           float64\n",
            " 413  palld                           float64\n",
            " 414  pbrm1                           float64\n",
            " 415  ppp2cb                          float64\n",
            " 416  ppp2r2a                         float64\n",
            " 417  prkacg                          float64\n",
            " 418  prkce                           float64\n",
            " 419  prkcq                           float64\n",
            " 420  prkcz                           float64\n",
            " 421  prkg1                           float64\n",
            " 422  prps2                           float64\n",
            " 423  prr16                           float64\n",
            " 424  ptpn22                          float64\n",
            " 425  ptprm                           float64\n",
            " 426  rasgef1b                        float64\n",
            " 427  rpgr                            float64\n",
            " 428  ryr2                            float64\n",
            " 429  sbno1                           float64\n",
            " 430  setd1a                          float64\n",
            " 431  setd2                           float64\n",
            " 432  setdb1                          float64\n",
            " 433  sf3b1                           float64\n",
            " 434  sgcd                            float64\n",
            " 435  shank2                          float64\n",
            " 436  siah1                           float64\n",
            " 437  sik1                            float64\n",
            " 438  sik2                            float64\n",
            " 439  smarcb1                         float64\n",
            " 440  smarcc1                         float64\n",
            " 441  smarcc2                         float64\n",
            " 442  smarcd1                         float64\n",
            " 443  spaca1                          float64\n",
            " 444  stab2                           float64\n",
            " 445  stmn2                           float64\n",
            " 446  syne1                           float64\n",
            " 447  taf1                            float64\n",
            " 448  taf4b                           float64\n",
            " 449  tbl1xr1                         float64\n",
            " 450  tg                              float64\n",
            " 451  thada                           float64\n",
            " 452  thsd7a                          float64\n",
            " 453  ttyh1                           float64\n",
            " 454  ubr5                            float64\n",
            " 455  ush2a                           float64\n",
            " 456  usp9x                           float64\n",
            " 457  utrn                            float64\n",
            " 458  zfp36l1                         float64\n",
            " 459  ackr3                           float64\n",
            " 460  akr1c1                          float64\n",
            " 461  akr1c2                          float64\n",
            " 462  akr1c3                          float64\n",
            " 463  akr1c4                          float64\n",
            " 464  akt3                            float64\n",
            " 465  ar                              float64\n",
            " 466  bche                            float64\n",
            " 467  cdk8                            float64\n",
            " 468  cdkn2c                          float64\n",
            " 469  cyb5a                           float64\n",
            " 470  cyp11a1                         float64\n",
            " 471  cyp11b2                         float64\n",
            " 472  cyp17a1                         float64\n",
            " 473  cyp19a1                         float64\n",
            " 474  cyp21a2                         float64\n",
            " 475  cyp3a43                         float64\n",
            " 476  cyp3a5                          float64\n",
            " 477  cyp3a7                          float64\n",
            " 478  ddc                             float64\n",
            " 479  hes6                            float64\n",
            " 480  hsd17b1                         float64\n",
            " 481  hsd17b10                        float64\n",
            " 482  hsd17b11                        float64\n",
            " 483  hsd17b12                        float64\n",
            " 484  hsd17b13                        float64\n",
            " 485  hsd17b14                        float64\n",
            " 486  hsd17b2                         float64\n",
            " 487  hsd17b3                         float64\n",
            " 488  hsd17b4                         float64\n",
            " 489  hsd17b6                         float64\n",
            " 490  hsd17b7                         float64\n",
            " 491  hsd17b8                         float64\n",
            " 492  hsd3b1                          float64\n",
            " 493  hsd3b2                          float64\n",
            " 494  hsd3b7                          float64\n",
            " 495  mecom                           float64\n",
            " 496  met                             float64\n",
            " 497  ncoa2                           float64\n",
            " 498  nrip1                           float64\n",
            " 499  pik3r3                          float64\n",
            " 500  prkci                           float64\n",
            " 501  prkd1                           float64\n",
            " 502  ran                             float64\n",
            " 503  rdh5                            float64\n",
            " 504  sdc4                            float64\n",
            " 505  serpini1                        float64\n",
            " 506  shbg                            float64\n",
            " 507  slc29a1                         float64\n",
            " 508  sox9                            float64\n",
            " 509  spry2                           float64\n",
            " 510  srd5a1                          float64\n",
            " 511  srd5a2                          float64\n",
            " 512  srd5a3                          float64\n",
            " 513  st7                             float64\n",
            " 514  star                            float64\n",
            " 515  tnk2                            float64\n",
            " 516  tulp4                           float64\n",
            " 517  ugt2b15                         float64\n",
            " 518  ugt2b17                         float64\n",
            " 519  ugt2b7                          float64\n",
            " 520  pik3ca_mut                      object \n",
            " 521  tp53_mut                        object \n",
            " 522  muc16_mut                       object \n",
            " 523  ahnak2_mut                      object \n",
            " 524  kmt2c_mut                       object \n",
            " 525  syne1_mut                       object \n",
            " 526  gata3_mut                       object \n",
            " 527  map3k1_mut                      object \n",
            " 528  ahnak_mut                       object \n",
            " 529  dnah11_mut                      object \n",
            " 530  cdh1_mut                        object \n",
            " 531  dnah2_mut                       object \n",
            " 532  kmt2d_mut                       object \n",
            " 533  ush2a_mut                       object \n",
            " 534  ryr2_mut                        object \n",
            " 535  dnah5_mut                       object \n",
            " 536  herc2_mut                       object \n",
            " 537  pde4dip_mut                     object \n",
            " 538  akap9_mut                       object \n",
            " 539  tg_mut                          object \n",
            " 540  birc6_mut                       object \n",
            " 541  utrn_mut                        object \n",
            " 542  tbx3_mut                        object \n",
            " 543  col6a3_mut                      object \n",
            " 544  arid1a_mut                      object \n",
            " 545  lama2_mut                       object \n",
            " 546  notch1_mut                      object \n",
            " 547  cbfb_mut                        object \n",
            " 548  ncor2_mut                       object \n",
            " 549  col12a1_mut                     object \n",
            " 550  col22a1_mut                     object \n",
            " 551  pten_mut                        object \n",
            " 552  akt1_mut                        object \n",
            " 553  atr_mut                         object \n",
            " 554  thada_mut                       object \n",
            " 555  ncor1_mut                       object \n",
            " 556  stab2_mut                       object \n",
            " 557  myh9_mut                        object \n",
            " 558  runx1_mut                       object \n",
            " 559  nf1_mut                         object \n",
            " 560  map2k4_mut                      object \n",
            " 561  ros1_mut                        object \n",
            " 562  lamb3_mut                       object \n",
            " 563  arid1b_mut                      object \n",
            " 564  erbb2_mut                       object \n",
            " 565  sf3b1_mut                       object \n",
            " 566  shank2_mut                      object \n",
            " 567  ep300_mut                       object \n",
            " 568  ptprd_mut                       object \n",
            " 569  usp9x_mut                       object \n",
            " 570  setd2_mut                       object \n",
            " 571  setd1a_mut                      object \n",
            " 572  thsd7a_mut                      object \n",
            " 573  afdn_mut                        object \n",
            " 574  erbb3_mut                       object \n",
            " 575  rb1_mut                         object \n",
            " 576  myo1a_mut                       object \n",
            " 577  alk_mut                         object \n",
            " 578  fanca_mut                       object \n",
            " 579  adgra2_mut                      object \n",
            " 580  ubr5_mut                        object \n",
            " 581  pik3r1_mut                      object \n",
            " 582  myo3a_mut                       object \n",
            " 583  asxl2_mut                       object \n",
            " 584  apc_mut                         object \n",
            " 585  ctcf_mut                        object \n",
            " 586  asxl1_mut                       object \n",
            " 587  fancd2_mut                      object \n",
            " 588  taf1_mut                        object \n",
            " 589  kdm6a_mut                       object \n",
            " 590  ctnna3_mut                      object \n",
            " 591  brca1_mut                       object \n",
            " 592  ptprm_mut                       object \n",
            " 593  foxo3_mut                       object \n",
            " 594  usp28_mut                       object \n",
            " 595  gldc_mut                        object \n",
            " 596  brca2_mut                       object \n",
            " 597  cacna2d3_mut                    object \n",
            " 598  arid2_mut                       object \n",
            " 599  aff2_mut                        object \n",
            " 600  lifr_mut                        object \n",
            " 601  sbno1_mut                       object \n",
            " 602  kdm3a_mut                       object \n",
            " 603  ncoa3_mut                       object \n",
            " 604  bap1_mut                        object \n",
            " 605  l1cam_mut                       object \n",
            " 606  pbrm1_mut                       object \n",
            " 607  chd1_mut                        object \n",
            " 608  jak1_mut                        object \n",
            " 609  setdb1_mut                      object \n",
            " 610  fam20c_mut                      object \n",
            " 611  arid5b_mut                      object \n",
            " 612  egfr_mut                        object \n",
            " 613  map3k10_mut                     object \n",
            " 614  smarcc2_mut                     object \n",
            " 615  erbb4_mut                       object \n",
            " 616  npnt_mut                        object \n",
            " 617  nek1_mut                        object \n",
            " 618  agmo_mut                        object \n",
            " 619  zfp36l1_mut                     object \n",
            " 620  smad4_mut                       object \n",
            " 621  sik1_mut                        object \n",
            " 622  casp8_mut                       object \n",
            " 623  prkcq_mut                       object \n",
            " 624  smarcc1_mut                     object \n",
            " 625  palld_mut                       object \n",
            " 626  dcaf4l2_mut                     object \n",
            " 627  bcas3_mut                       object \n",
            " 628  cdkn1b_mut                      object \n",
            " 629  gps2_mut                        object \n",
            " 630  men1_mut                        object \n",
            " 631  stk11_mut                       object \n",
            " 632  sik2_mut                        object \n",
            " 633  ptpn22_mut                      object \n",
            " 634  brip1_mut                       object \n",
            " 635  flt3_mut                        object \n",
            " 636  nrg3_mut                        object \n",
            " 637  fbxw7_mut                       object \n",
            " 638  ttyh1_mut                       object \n",
            " 639  taf4b_mut                       object \n",
            " 640  or6a2_mut                       object \n",
            " 641  map3k13_mut                     object \n",
            " 642  hdac9_mut                       object \n",
            " 643  prkacg_mut                      object \n",
            " 644  rpgr_mut                        object \n",
            " 645  large1_mut                      object \n",
            " 646  foxp1_mut                       object \n",
            " 647  clk3_mut                        object \n",
            " 648  prkcz_mut                       object \n",
            " 649  lipi_mut                        object \n",
            " 650  ppp2r2a_mut                     object \n",
            " 651  prkce_mut                       object \n",
            " 652  gh1_mut                         object \n",
            " 653  gpr32_mut                       object \n",
            " 654  kras_mut                        object \n",
            " 655  nf2_mut                         object \n",
            " 656  chek2_mut                       object \n",
            " 657  ldlrap1_mut                     object \n",
            " 658  clrn2_mut                       object \n",
            " 659  acvrl1_mut                      object \n",
            " 660  agtr2_mut                       float64\n",
            " 661  cdkn2a_mut                      object \n",
            " 662  ctnna1_mut                      object \n",
            " 663  magea8_mut                      object \n",
            " 664  prr16_mut                       object \n",
            " 665  dtwd2_mut                       object \n",
            " 666  akt2_mut                        object \n",
            " 667  braf_mut                        object \n",
            " 668  foxo1_mut                       object \n",
            " 669  nt5e_mut                        object \n",
            " 670  ccnd3_mut                       float64\n",
            " 671  nr3c1_mut                       object \n",
            " 672  prkg1_mut                       object \n",
            " 673  tbl1xr1_mut                     object \n",
            " 674  frmd3_mut                       object \n",
            " 675  smad2_mut                       float64\n",
            " 676  sgcd_mut                        object \n",
            " 677  spaca1_mut                      object \n",
            " 678  rasgef1b_mut                    float64\n",
            " 679  hist1h2bc_mut                   object \n",
            " 680  nr2f1_mut                       object \n",
            " 681  klrg1_mut                       object \n",
            " 682  mbl2_mut                        float64\n",
            " 683  mtap_mut                        object \n",
            " 684  ppp2cb_mut                      object \n",
            " 685  smarcd1_mut                     object \n",
            " 686  nras_mut                        float64\n",
            " 687  ndfip1_mut                      object \n",
            " 688  hras_mut                        float64\n",
            " 689  prps2_mut                       object \n",
            " 690  smarcb1_mut                     float64\n",
            " 691  stmn2_mut                       float64\n",
            " 692  siah1_mut                       float64\n",
            "dtypes: float64(508), int64(5), object(180)\n",
            "memory usage: 3.8+ MB\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "columns = data.columns"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "l473UNp5pC69",
        "outputId": "b95516e0-f90e-40c1-af76-8802d65699b3"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['patient_id', 'age_at_diagnosis', 'type_of_breast_surgery',\n",
              "       'cancer_type', 'cancer_type_detailed', 'cellularity', 'chemotherapy',\n",
              "       'pam50_+_claudin-low_subtype', 'cohort', 'er_status_measured_by_ihc',\n",
              "       ...\n",
              "       'mtap_mut', 'ppp2cb_mut', 'smarcd1_mut', 'nras_mut', 'ndfip1_mut',\n",
              "       'hras_mut', 'prps2_mut', 'smarcb1_mut', 'stmn2_mut', 'siah1_mut'],\n",
              "      dtype='object', length=693)"
            ]
          },
          "metadata": {},
          "execution_count": 3
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "for i, column in enumerate(columns, start=1):\n",
        "  print(f\"{i}: {column} has unique values: {data[column].nunique()}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "KwXDBK1_pLCg",
        "outputId": "90d72ddc-e663-4092-f663-15f12f16ff1d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "1: patient_id has unique values: 715\n",
            "2: age_at_diagnosis has unique values: 664\n",
            "3: type_of_breast_surgery has unique values: 2\n",
            "4: cancer_type has unique values: 2\n",
            "5: cancer_type_detailed has unique values: 6\n",
            "6: cellularity has unique values: 3\n",
            "7: chemotherapy has unique values: 2\n",
            "8: pam50_+_claudin-low_subtype has unique values: 7\n",
            "9: cohort has unique values: 2\n",
            "10: er_status_measured_by_ihc has unique values: 2\n",
            "11: er_status has unique values: 2\n",
            "12: neoplasm_histologic_grade has unique values: 3\n",
            "13: her2_status_measured_by_snp6 has unique values: 3\n",
            "14: her2_status has unique values: 2\n",
            "15: tumor_other_histologic_subtype has unique values: 8\n",
            "16: hormone_therapy has unique values: 2\n",
            "17: inferred_menopausal_state has unique values: 2\n",
            "18: integrative_cluster has unique values: 11\n",
            "19: primary_tumor_laterality has unique values: 2\n",
            "20: lymph_nodes_examined_positive has unique values: 27\n",
            "21: mutation_count has unique values: 19\n",
            "22: nottingham_prognostic_index has unique values: 198\n",
            "23: oncotree_code has unique values: 6\n",
            "24: overall_survival_months has unique values: 678\n",
            "25: overall_survival has unique values: 2\n",
            "26: pr_status has unique values: 2\n",
            "27: radio_therapy has unique values: 2\n",
            "28: 3-gene_classifier_subtype has unique values: 4\n",
            "29: tumor_size has unique values: 59\n",
            "30: tumor_stage has unique values: 5\n",
            "31: death_from_cancer has unique values: 3\n",
            "32: brca1 has unique values: 705\n",
            "33: brca2 has unique values: 706\n",
            "34: palb2 has unique values: 708\n",
            "35: pten has unique values: 708\n",
            "36: tp53 has unique values: 707\n",
            "37: atm has unique values: 709\n",
            "38: cdh1 has unique values: 706\n",
            "39: chek2 has unique values: 695\n",
            "40: nbn has unique values: 703\n",
            "41: nf1 has unique values: 695\n",
            "42: stk11 has unique values: 693\n",
            "43: bard1 has unique values: 707\n",
            "44: mlh1 has unique values: 707\n",
            "45: msh2 has unique values: 708\n",
            "46: msh6 has unique values: 709\n",
            "47: pms2 has unique values: 699\n",
            "48: epcam has unique values: 710\n",
            "49: rad51c has unique values: 693\n",
            "50: rad51d has unique values: 698\n",
            "51: rad50 has unique values: 705\n",
            "52: rb1 has unique values: 703\n",
            "53: rbl1 has unique values: 699\n",
            "54: rbl2 has unique values: 695\n",
            "55: ccna1 has unique values: 696\n",
            "56: ccnb1 has unique values: 709\n",
            "57: cdk1 has unique values: 709\n",
            "58: ccne1 has unique values: 705\n",
            "59: cdk2 has unique values: 704\n",
            "60: cdc25a has unique values: 701\n",
            "61: ccnd1 has unique values: 598\n",
            "62: cdk4 has unique values: 661\n",
            "63: cdk6 has unique values: 707\n",
            "64: ccnd2 has unique values: 708\n",
            "65: cdkn2a has unique values: 703\n",
            "66: cdkn2b has unique values: 710\n",
            "67: myc has unique values: 707\n",
            "68: cdkn1a has unique values: 706\n",
            "69: cdkn1b has unique values: 708\n",
            "70: e2f1 has unique values: 705\n",
            "71: e2f2 has unique values: 708\n",
            "72: e2f3 has unique values: 709\n",
            "73: e2f4 has unique values: 688\n",
            "74: e2f5 has unique values: 703\n",
            "75: e2f6 has unique values: 705\n",
            "76: e2f7 has unique values: 711\n",
            "77: e2f8 has unique values: 704\n",
            "78: src has unique values: 691\n",
            "79: jak1 has unique values: 698\n",
            "80: jak2 has unique values: 701\n",
            "81: stat1 has unique values: 686\n",
            "82: stat2 has unique values: 662\n",
            "83: stat3 has unique values: 692\n",
            "84: stat5a has unique values: 698\n",
            "85: stat5b has unique values: 704\n",
            "86: mdm2 has unique values: 709\n",
            "87: tp53bp1 has unique values: 706\n",
            "88: adam10 has unique values: 698\n",
            "89: adam17 has unique values: 698\n",
            "90: aph1a has unique values: 697\n",
            "91: aph1b has unique values: 707\n",
            "92: arrdc1 has unique values: 699\n",
            "93: cir1 has unique values: 698\n",
            "94: ctbp1 has unique values: 680\n",
            "95: ctbp2 has unique values: 710\n",
            "96: cul1 has unique values: 685\n",
            "97: dll1 has unique values: 703\n",
            "98: dll3 has unique values: 696\n",
            "99: dll4 has unique values: 704\n",
            "100: dtx1 has unique values: 705\n",
            "101: dtx2 has unique values: 704\n",
            "102: dtx3 has unique values: 703\n",
            "103: dtx4 has unique values: 706\n",
            "104: ep300 has unique values: 703\n",
            "105: fbxw7 has unique values: 697\n",
            "106: hdac1 has unique values: 644\n",
            "107: hdac2 has unique values: 687\n",
            "108: hes1 has unique values: 695\n",
            "109: hes5 has unique values: 701\n",
            "110: heyl has unique values: 695\n",
            "111: itch has unique values: 689\n",
            "112: jag1 has unique values: 696\n",
            "113: jag2 has unique values: 702\n",
            "114: kdm5a has unique values: 708\n",
            "115: lfng has unique values: 698\n",
            "116: maml1 has unique values: 709\n",
            "117: maml2 has unique values: 705\n",
            "118: maml3 has unique values: 697\n",
            "119: ncor2 has unique values: 708\n",
            "120: ncstn has unique values: 709\n",
            "121: notch1 has unique values: 702\n",
            "122: notch2 has unique values: 698\n",
            "123: notch3 has unique values: 708\n",
            "124: nrarp has unique values: 704\n",
            "125: numb has unique values: 706\n",
            "126: numbl has unique values: 704\n",
            "127: psen1 has unique values: 705\n",
            "128: psen2 has unique values: 704\n",
            "129: psenen has unique values: 695\n",
            "130: rbpj has unique values: 710\n",
            "131: rbpjl has unique values: 702\n",
            "132: rfng has unique values: 709\n",
            "133: snw1 has unique values: 704\n",
            "134: spen has unique values: 685\n",
            "135: hes2 has unique values: 704\n",
            "136: hes4 has unique values: 674\n",
            "137: hes7 has unique values: 651\n",
            "138: hey1 has unique values: 694\n",
            "139: hey2 has unique values: 705\n",
            "140: acvr1 has unique values: 687\n",
            "141: acvr1b has unique values: 690\n",
            "142: acvr1c has unique values: 708\n",
            "143: acvr2a has unique values: 700\n",
            "144: acvr2b has unique values: 707\n",
            "145: acvrl1 has unique values: 699\n",
            "146: akt1 has unique values: 706\n",
            "147: akt1s1 has unique values: 704\n",
            "148: akt2 has unique values: 704\n",
            "149: apaf1 has unique values: 687\n",
            "150: arl11 has unique values: 704\n",
            "151: atr has unique values: 706\n",
            "152: aurka has unique values: 711\n",
            "153: bad has unique values: 709\n",
            "154: bcl2 has unique values: 711\n",
            "155: bcl2l1 has unique values: 649\n",
            "156: bmp10 has unique values: 707\n",
            "157: bmp15 has unique values: 711\n",
            "158: bmp2 has unique values: 702\n",
            "159: bmp3 has unique values: 709\n",
            "160: bmp4 has unique values: 706\n",
            "161: bmp5 has unique values: 702\n",
            "162: bmp6 has unique values: 698\n",
            "163: bmp7 has unique values: 700\n",
            "164: bmpr1a has unique values: 703\n",
            "165: bmpr1b has unique values: 702\n",
            "166: bmpr2 has unique values: 691\n",
            "167: braf has unique values: 708\n",
            "168: casp10 has unique values: 712\n",
            "169: casp3 has unique values: 682\n",
            "170: casp6 has unique values: 698\n",
            "171: casp7 has unique values: 708\n",
            "172: casp8 has unique values: 705\n",
            "173: casp9 has unique values: 708\n",
            "174: chek1 has unique values: 700\n",
            "175: csf1 has unique values: 707\n",
            "176: csf1r has unique values: 672\n",
            "177: cxcl8 has unique values: 706\n",
            "178: cxcr1 has unique values: 703\n",
            "179: cxcr2 has unique values: 709\n",
            "180: dab2 has unique values: 685\n",
            "181: diras3 has unique values: 702\n",
            "182: dlec1 has unique values: 708\n",
            "183: dph1 has unique values: 707\n",
            "184: egfr has unique values: 703\n",
            "185: eif4e has unique values: 713\n",
            "186: eif4ebp1 has unique values: 704\n",
            "187: eif5a2 has unique values: 698\n",
            "188: erbb2 has unique values: 704\n",
            "189: erbb3 has unique values: 693\n",
            "190: erbb4 has unique values: 700\n",
            "191: fas has unique values: 709\n",
            "192: fgf1 has unique values: 707\n",
            "193: fgfr1 has unique values: 710\n",
            "194: folr1 has unique values: 708\n",
            "195: folr2 has unique values: 701\n",
            "196: folr3 has unique values: 703\n",
            "197: foxo1 has unique values: 698\n",
            "198: foxo3 has unique values: 705\n",
            "199: gdf11 has unique values: 695\n",
            "200: gdf2 has unique values: 705\n",
            "201: gsk3b has unique values: 699\n",
            "202: hif1a has unique values: 712\n",
            "203: hla-g has unique values: 696\n",
            "204: hras has unique values: 706\n",
            "205: igf1 has unique values: 701\n",
            "206: igf1r has unique values: 707\n",
            "207: inha has unique values: 696\n",
            "208: inhba has unique values: 705\n",
            "209: inhbc has unique values: 701\n",
            "210: itgav has unique values: 696\n",
            "211: itgb3 has unique values: 698\n",
            "212: izumo1r has unique values: 701\n",
            "213: kdr has unique values: 710\n",
            "214: kit has unique values: 708\n",
            "215: kras has unique values: 696\n",
            "216: map2k1 has unique values: 707\n",
            "217: map2k2 has unique values: 698\n",
            "218: map2k3 has unique values: 707\n",
            "219: map2k4 has unique values: 708\n",
            "220: map2k5 has unique values: 695\n",
            "221: map3k1 has unique values: 699\n",
            "222: map3k3 has unique values: 694\n",
            "223: map3k4 has unique values: 693\n",
            "224: map3k5 has unique values: 701\n",
            "225: mapk1 has unique values: 700\n",
            "226: mapk12 has unique values: 708\n",
            "227: mapk14 has unique values: 701\n",
            "228: mapk3 has unique values: 687\n",
            "229: mapk4 has unique values: 702\n",
            "230: mapk6 has unique values: 699\n",
            "231: mapk7 has unique values: 698\n",
            "232: mapk8 has unique values: 708\n",
            "233: mapk9 has unique values: 697\n",
            "234: mdc1 has unique values: 682\n",
            "235: mlst8 has unique values: 694\n",
            "236: mmp1 has unique values: 698\n",
            "237: mmp10 has unique values: 699\n",
            "238: mmp11 has unique values: 705\n",
            "239: mmp12 has unique values: 698\n",
            "240: mmp13 has unique values: 707\n",
            "241: mmp14 has unique values: 705\n",
            "242: mmp15 has unique values: 696\n",
            "243: mmp16 has unique values: 703\n",
            "244: mmp17 has unique values: 701\n",
            "245: mmp19 has unique values: 704\n",
            "246: mmp2 has unique values: 711\n",
            "247: mmp21 has unique values: 706\n",
            "248: mmp23b has unique values: 706\n",
            "249: mmp24 has unique values: 702\n",
            "250: mmp25 has unique values: 706\n",
            "251: mmp26 has unique values: 701\n",
            "252: mmp27 has unique values: 705\n",
            "253: mmp28 has unique values: 703\n",
            "254: mmp3 has unique values: 703\n",
            "255: mmp7 has unique values: 698\n",
            "256: mmp9 has unique values: 702\n",
            "257: mtor has unique values: 683\n",
            "258: nfkb1 has unique values: 642\n",
            "259: nfkb2 has unique values: 694\n",
            "260: opcml has unique values: 706\n",
            "261: pdgfa has unique values: 705\n",
            "262: pdgfb has unique values: 702\n",
            "263: pdgfra has unique values: 707\n",
            "264: pdgfrb has unique values: 681\n",
            "265: pdpk1 has unique values: 699\n",
            "266: peg3 has unique values: 706\n",
            "267: pik3ca has unique values: 705\n",
            "268: pik3r1 has unique values: 696\n",
            "269: pik3r2 has unique values: 663\n",
            "270: plagl1 has unique values: 708\n",
            "271: ptk2 has unique values: 710\n",
            "272: rab25 has unique values: 673\n",
            "273: rad51 has unique values: 703\n",
            "274: raf1 has unique values: 693\n",
            "275: rassf1 has unique values: 692\n",
            "276: rheb has unique values: 702\n",
            "277: rictor has unique values: 703\n",
            "278: rps6 has unique values: 711\n",
            "279: rps6ka1 has unique values: 691\n",
            "280: rps6ka2 has unique values: 697\n",
            "281: rps6kb1 has unique values: 706\n",
            "282: rps6kb2 has unique values: 704\n",
            "283: rptor has unique values: 706\n",
            "284: slc19a1 has unique values: 706\n",
            "285: smad1 has unique values: 708\n",
            "286: smad2 has unique values: 702\n",
            "287: smad3 has unique values: 691\n",
            "288: smad4 has unique values: 688\n",
            "289: smad5 has unique values: 698\n",
            "290: smad6 has unique values: 706\n",
            "291: smad7 has unique values: 709\n",
            "292: smad9 has unique values: 697\n",
            "293: sptbn1 has unique values: 692\n",
            "294: terc has unique values: 703\n",
            "295: tert has unique values: 708\n",
            "296: tgfb1 has unique values: 696\n",
            "297: tgfb2 has unique values: 701\n",
            "298: tgfb3 has unique values: 709\n",
            "299: tgfbr1 has unique values: 705\n",
            "300: tgfbr2 has unique values: 700\n",
            "301: tgfbr3 has unique values: 700\n",
            "302: tsc1 has unique values: 695\n",
            "303: tsc2 has unique values: 710\n",
            "304: vegfa has unique values: 697\n",
            "305: vegfb has unique values: 615\n",
            "306: wfdc2 has unique values: 705\n",
            "307: wwox has unique values: 694\n",
            "308: zfyve9 has unique values: 709\n",
            "309: arid1a has unique values: 685\n",
            "310: arid1b has unique values: 709\n",
            "311: cbfb has unique values: 684\n",
            "312: gata3 has unique values: 707\n",
            "313: kmt2c has unique values: 708\n",
            "314: kmt2d has unique values: 699\n",
            "315: myh9 has unique values: 701\n",
            "316: ncor1 has unique values: 705\n",
            "317: pde4dip has unique values: 707\n",
            "318: ptprd has unique values: 701\n",
            "319: ros1 has unique values: 699\n",
            "320: runx1 has unique values: 698\n",
            "321: tbx3 has unique values: 706\n",
            "322: abcb1 has unique values: 705\n",
            "323: abcb11 has unique values: 700\n",
            "324: abcc1 has unique values: 701\n",
            "325: abcc10 has unique values: 696\n",
            "326: bbc3 has unique values: 705\n",
            "327: bmf has unique values: 696\n",
            "328: cyp2c8 has unique values: 695\n",
            "329: cyp3a4 has unique values: 697\n",
            "330: fgf2 has unique values: 707\n",
            "331: fn1 has unique values: 710\n",
            "332: map2 has unique values: 700\n",
            "333: map4 has unique values: 702\n",
            "334: mapt has unique values: 710\n",
            "335: nr1i2 has unique values: 712\n",
            "336: slco1b3 has unique values: 703\n",
            "337: tubb1 has unique values: 707\n",
            "338: tubb4a has unique values: 694\n",
            "339: tubb4b has unique values: 705\n",
            "340: twist1 has unique values: 708\n",
            "341: adgra2 has unique values: 706\n",
            "342: afdn has unique values: 686\n",
            "343: aff2 has unique values: 705\n",
            "344: agmo has unique values: 703\n",
            "345: agtr2 has unique values: 706\n",
            "346: ahnak has unique values: 584\n",
            "347: ahnak2 has unique values: 702\n",
            "348: akap9 has unique values: 708\n",
            "349: alk has unique values: 701\n",
            "350: apc has unique values: 705\n",
            "351: arid2 has unique values: 696\n",
            "352: arid5b has unique values: 686\n",
            "353: asxl1 has unique values: 695\n",
            "354: asxl2 has unique values: 680\n",
            "355: bap1 has unique values: 691\n",
            "356: bcas3 has unique values: 694\n",
            "357: birc6 has unique values: 693\n",
            "358: cacna2d3 has unique values: 703\n",
            "359: ccnd3 has unique values: 650\n",
            "360: chd1 has unique values: 699\n",
            "361: clk3 has unique values: 708\n",
            "362: clrn2 has unique values: 704\n",
            "363: col12a1 has unique values: 706\n",
            "364: col22a1 has unique values: 695\n",
            "365: col6a3 has unique values: 644\n",
            "366: ctcf has unique values: 693\n",
            "367: ctnna1 has unique values: 705\n",
            "368: ctnna3 has unique values: 699\n",
            "369: dnah11 has unique values: 695\n",
            "370: dnah2 has unique values: 710\n",
            "371: dnah5 has unique values: 700\n",
            "372: dtwd2 has unique values: 703\n",
            "373: fam20c has unique values: 705\n",
            "374: fanca has unique values: 708\n",
            "375: fancd2 has unique values: 705\n",
            "376: flt3 has unique values: 702\n",
            "377: foxp1 has unique values: 703\n",
            "378: frmd3 has unique values: 698\n",
            "379: gh1 has unique values: 707\n",
            "380: gldc has unique values: 692\n",
            "381: gpr32 has unique values: 708\n",
            "382: gps2 has unique values: 705\n",
            "383: hdac9 has unique values: 701\n",
            "384: herc2 has unique values: 686\n",
            "385: hist1h2bc has unique values: 709\n",
            "386: kdm3a has unique values: 696\n",
            "387: kdm6a has unique values: 704\n",
            "388: klrg1 has unique values: 699\n",
            "389: l1cam has unique values: 701\n",
            "390: lama2 has unique values: 710\n",
            "391: lamb3 has unique values: 706\n",
            "392: large1 has unique values: 696\n",
            "393: ldlrap1 has unique values: 695\n",
            "394: lifr has unique values: 708\n",
            "395: lipi has unique values: 700\n",
            "396: magea8 has unique values: 694\n",
            "397: map3k10 has unique values: 708\n",
            "398: map3k13 has unique values: 703\n",
            "399: men1 has unique values: 711\n",
            "400: mtap has unique values: 688\n",
            "401: muc16 has unique values: 695\n",
            "402: myo1a has unique values: 708\n",
            "403: myo3a has unique values: 703\n",
            "404: ncoa3 has unique values: 687\n",
            "405: nek1 has unique values: 698\n",
            "406: nf2 has unique values: 701\n",
            "407: npnt has unique values: 700\n",
            "408: nr2f1 has unique values: 705\n",
            "409: nr3c1 has unique values: 704\n",
            "410: nras has unique values: 693\n",
            "411: nrg3 has unique values: 702\n",
            "412: nt5e has unique values: 706\n",
            "413: or6a2 has unique values: 707\n",
            "414: palld has unique values: 705\n",
            "415: pbrm1 has unique values: 699\n",
            "416: ppp2cb has unique values: 704\n",
            "417: ppp2r2a has unique values: 705\n",
            "418: prkacg has unique values: 705\n",
            "419: prkce has unique values: 702\n",
            "420: prkcq has unique values: 694\n",
            "421: prkcz has unique values: 702\n",
            "422: prkg1 has unique values: 700\n",
            "423: prps2 has unique values: 709\n",
            "424: prr16 has unique values: 696\n",
            "425: ptpn22 has unique values: 697\n",
            "426: ptprm has unique values: 704\n",
            "427: rasgef1b has unique values: 709\n",
            "428: rpgr has unique values: 704\n",
            "429: ryr2 has unique values: 701\n",
            "430: sbno1 has unique values: 702\n",
            "431: setd1a has unique values: 695\n",
            "432: setd2 has unique values: 676\n",
            "433: setdb1 has unique values: 696\n",
            "434: sf3b1 has unique values: 703\n",
            "435: sgcd has unique values: 705\n",
            "436: shank2 has unique values: 703\n",
            "437: siah1 has unique values: 708\n",
            "438: sik1 has unique values: 700\n",
            "439: sik2 has unique values: 699\n",
            "440: smarcb1 has unique values: 698\n",
            "441: smarcc1 has unique values: 687\n",
            "442: smarcc2 has unique values: 691\n",
            "443: smarcd1 has unique values: 676\n",
            "444: spaca1 has unique values: 702\n",
            "445: stab2 has unique values: 703\n",
            "446: stmn2 has unique values: 702\n",
            "447: syne1 has unique values: 699\n",
            "448: taf1 has unique values: 710\n",
            "449: taf4b has unique values: 704\n",
            "450: tbl1xr1 has unique values: 708\n",
            "451: tg has unique values: 703\n",
            "452: thada has unique values: 689\n",
            "453: thsd7a has unique values: 702\n",
            "454: ttyh1 has unique values: 695\n",
            "455: ubr5 has unique values: 692\n",
            "456: ush2a has unique values: 701\n",
            "457: usp9x has unique values: 707\n",
            "458: utrn has unique values: 697\n",
            "459: zfp36l1 has unique values: 698\n",
            "460: ackr3 has unique values: 701\n",
            "461: akr1c1 has unique values: 706\n",
            "462: akr1c2 has unique values: 706\n",
            "463: akr1c3 has unique values: 698\n",
            "464: akr1c4 has unique values: 707\n",
            "465: akt3 has unique values: 702\n",
            "466: ar has unique values: 705\n",
            "467: bche has unique values: 697\n",
            "468: cdk8 has unique values: 708\n",
            "469: cdkn2c has unique values: 709\n",
            "470: cyb5a has unique values: 704\n",
            "471: cyp11a1 has unique values: 698\n",
            "472: cyp11b2 has unique values: 700\n",
            "473: cyp17a1 has unique values: 703\n",
            "474: cyp19a1 has unique values: 709\n",
            "475: cyp21a2 has unique values: 703\n",
            "476: cyp3a43 has unique values: 699\n",
            "477: cyp3a5 has unique values: 701\n",
            "478: cyp3a7 has unique values: 706\n",
            "479: ddc has unique values: 684\n",
            "480: hes6 has unique values: 705\n",
            "481: hsd17b1 has unique values: 705\n",
            "482: hsd17b10 has unique values: 706\n",
            "483: hsd17b11 has unique values: 704\n",
            "484: hsd17b12 has unique values: 705\n",
            "485: hsd17b13 has unique values: 708\n",
            "486: hsd17b14 has unique values: 696\n",
            "487: hsd17b2 has unique values: 699\n",
            "488: hsd17b3 has unique values: 709\n",
            "489: hsd17b4 has unique values: 703\n",
            "490: hsd17b6 has unique values: 702\n",
            "491: hsd17b7 has unique values: 706\n",
            "492: hsd17b8 has unique values: 707\n",
            "493: hsd3b1 has unique values: 710\n",
            "494: hsd3b2 has unique values: 709\n",
            "495: hsd3b7 has unique values: 695\n",
            "496: mecom has unique values: 705\n",
            "497: met has unique values: 707\n",
            "498: ncoa2 has unique values: 704\n",
            "499: nrip1 has unique values: 695\n",
            "500: pik3r3 has unique values: 699\n",
            "501: prkci has unique values: 706\n",
            "502: prkd1 has unique values: 710\n",
            "503: ran has unique values: 702\n",
            "504: rdh5 has unique values: 709\n",
            "505: sdc4 has unique values: 707\n",
            "506: serpini1 has unique values: 706\n",
            "507: shbg has unique values: 703\n",
            "508: slc29a1 has unique values: 700\n",
            "509: sox9 has unique values: 701\n",
            "510: spry2 has unique values: 702\n",
            "511: srd5a1 has unique values: 706\n",
            "512: srd5a2 has unique values: 708\n",
            "513: srd5a3 has unique values: 702\n",
            "514: st7 has unique values: 710\n",
            "515: star has unique values: 708\n",
            "516: tnk2 has unique values: 691\n",
            "517: tulp4 has unique values: 687\n",
            "518: ugt2b15 has unique values: 706\n",
            "519: ugt2b17 has unique values: 689\n",
            "520: ugt2b7 has unique values: 695\n",
            "521: pik3ca_mut has unique values: 72\n",
            "522: tp53_mut has unique values: 156\n",
            "523: muc16_mut has unique values: 77\n",
            "524: ahnak2_mut has unique values: 57\n",
            "525: kmt2c_mut has unique values: 57\n",
            "526: syne1_mut has unique values: 62\n",
            "527: gata3_mut has unique values: 64\n",
            "528: map3k1_mut has unique values: 69\n",
            "529: ahnak_mut has unique values: 39\n",
            "530: dnah11_mut has unique values: 41\n",
            "531: cdh1_mut has unique values: 56\n",
            "532: dnah2_mut has unique values: 36\n",
            "533: kmt2d_mut has unique values: 36\n",
            "534: ush2a_mut has unique values: 40\n",
            "535: ryr2_mut has unique values: 40\n",
            "536: dnah5_mut has unique values: 33\n",
            "537: herc2_mut has unique values: 28\n",
            "538: pde4dip_mut has unique values: 20\n",
            "539: akap9_mut has unique values: 26\n",
            "540: tg_mut has unique values: 18\n",
            "541: birc6_mut has unique values: 28\n",
            "542: utrn_mut has unique values: 26\n",
            "543: tbx3_mut has unique values: 41\n",
            "544: col6a3_mut has unique values: 16\n",
            "545: arid1a_mut has unique values: 32\n",
            "546: lama2_mut has unique values: 27\n",
            "547: notch1_mut has unique values: 20\n",
            "548: cbfb_mut has unique values: 32\n",
            "549: ncor2_mut has unique values: 28\n",
            "550: col12a1_mut has unique values: 16\n",
            "551: col22a1_mut has unique values: 21\n",
            "552: pten_mut has unique values: 38\n",
            "553: akt1_mut has unique values: 4\n",
            "554: atr_mut has unique values: 27\n",
            "555: thada_mut has unique values: 17\n",
            "556: ncor1_mut has unique values: 26\n",
            "557: stab2_mut has unique values: 19\n",
            "558: myh9_mut has unique values: 25\n",
            "559: runx1_mut has unique values: 21\n",
            "560: nf1_mut has unique values: 20\n",
            "561: map2k4_mut has unique values: 23\n",
            "562: ros1_mut has unique values: 16\n",
            "563: lamb3_mut has unique values: 18\n",
            "564: arid1b_mut has unique values: 11\n",
            "565: erbb2_mut has unique values: 16\n",
            "566: sf3b1_mut has unique values: 6\n",
            "567: shank2_mut has unique values: 14\n",
            "568: ep300_mut has unique values: 9\n",
            "569: ptprd_mut has unique values: 15\n",
            "570: usp9x_mut has unique values: 11\n",
            "571: setd2_mut has unique values: 16\n",
            "572: setd1a_mut has unique values: 13\n",
            "573: thsd7a_mut has unique values: 14\n",
            "574: afdn_mut has unique values: 13\n",
            "575: erbb3_mut has unique values: 17\n",
            "576: rb1_mut has unique values: 16\n",
            "577: myo1a_mut has unique values: 9\n",
            "578: alk_mut has unique values: 18\n",
            "579: fanca_mut has unique values: 9\n",
            "580: adgra2_mut has unique values: 11\n",
            "581: ubr5_mut has unique values: 14\n",
            "582: pik3r1_mut has unique values: 10\n",
            "583: myo3a_mut has unique values: 12\n",
            "584: asxl2_mut has unique values: 14\n",
            "585: apc_mut has unique values: 9\n",
            "586: ctcf_mut has unique values: 16\n",
            "587: asxl1_mut has unique values: 10\n",
            "588: fancd2_mut has unique values: 14\n",
            "589: taf1_mut has unique values: 14\n",
            "590: kdm6a_mut has unique values: 17\n",
            "591: ctnna3_mut has unique values: 5\n",
            "592: brca1_mut has unique values: 9\n",
            "593: ptprm_mut has unique values: 8\n",
            "594: foxo3_mut has unique values: 12\n",
            "595: usp28_mut has unique values: 12\n",
            "596: gldc_mut has unique values: 9\n",
            "597: brca2_mut has unique values: 11\n",
            "598: cacna2d3_mut has unique values: 4\n",
            "599: arid2_mut has unique values: 10\n",
            "600: aff2_mut has unique values: 9\n",
            "601: lifr_mut has unique values: 8\n",
            "602: sbno1_mut has unique values: 9\n",
            "603: kdm3a_mut has unique values: 5\n",
            "604: ncoa3_mut has unique values: 7\n",
            "605: bap1_mut has unique values: 13\n",
            "606: l1cam_mut has unique values: 12\n",
            "607: pbrm1_mut has unique values: 9\n",
            "608: chd1_mut has unique values: 10\n",
            "609: jak1_mut has unique values: 8\n",
            "610: setdb1_mut has unique values: 6\n",
            "611: fam20c_mut has unique values: 6\n",
            "612: arid5b_mut has unique values: 6\n",
            "613: egfr_mut has unique values: 5\n",
            "614: map3k10_mut has unique values: 7\n",
            "615: smarcc2_mut has unique values: 7\n",
            "616: erbb4_mut has unique values: 5\n",
            "617: npnt_mut has unique values: 8\n",
            "618: nek1_mut has unique values: 8\n",
            "619: agmo_mut has unique values: 5\n",
            "620: zfp36l1_mut has unique values: 10\n",
            "621: smad4_mut has unique values: 3\n",
            "622: sik1_mut has unique values: 5\n",
            "623: casp8_mut has unique values: 11\n",
            "624: prkcq_mut has unique values: 8\n",
            "625: smarcc1_mut has unique values: 7\n",
            "626: palld_mut has unique values: 8\n",
            "627: dcaf4l2_mut has unique values: 6\n",
            "628: bcas3_mut has unique values: 9\n",
            "629: cdkn1b_mut has unique values: 9\n",
            "630: gps2_mut has unique values: 7\n",
            "631: men1_mut has unique values: 11\n",
            "632: stk11_mut has unique values: 6\n",
            "633: sik2_mut has unique values: 5\n",
            "634: ptpn22_mut has unique values: 2\n",
            "635: brip1_mut has unique values: 9\n",
            "636: flt3_mut has unique values: 8\n",
            "637: nrg3_mut has unique values: 6\n",
            "638: fbxw7_mut has unique values: 9\n",
            "639: ttyh1_mut has unique values: 7\n",
            "640: taf4b_mut has unique values: 5\n",
            "641: or6a2_mut has unique values: 4\n",
            "642: map3k13_mut has unique values: 9\n",
            "643: hdac9_mut has unique values: 9\n",
            "644: prkacg_mut has unique values: 4\n",
            "645: rpgr_mut has unique values: 5\n",
            "646: large1_mut has unique values: 4\n",
            "647: foxp1_mut has unique values: 7\n",
            "648: clk3_mut has unique values: 7\n",
            "649: prkcz_mut has unique values: 3\n",
            "650: lipi_mut has unique values: 3\n",
            "651: ppp2r2a_mut has unique values: 4\n",
            "652: prkce_mut has unique values: 4\n",
            "653: gh1_mut has unique values: 4\n",
            "654: gpr32_mut has unique values: 5\n",
            "655: kras_mut has unique values: 3\n",
            "656: nf2_mut has unique values: 6\n",
            "657: chek2_mut has unique values: 7\n",
            "658: ldlrap1_mut has unique values: 3\n",
            "659: clrn2_mut has unique values: 2\n",
            "660: acvrl1_mut has unique values: 5\n",
            "661: agtr2_mut has unique values: 1\n",
            "662: cdkn2a_mut has unique values: 6\n",
            "663: ctnna1_mut has unique values: 2\n",
            "664: magea8_mut has unique values: 3\n",
            "665: prr16_mut has unique values: 2\n",
            "666: dtwd2_mut has unique values: 4\n",
            "667: akt2_mut has unique values: 5\n",
            "668: braf_mut has unique values: 3\n",
            "669: foxo1_mut has unique values: 6\n",
            "670: nt5e_mut has unique values: 2\n",
            "671: ccnd3_mut has unique values: 1\n",
            "672: nr3c1_mut has unique values: 2\n",
            "673: prkg1_mut has unique values: 2\n",
            "674: tbl1xr1_mut has unique values: 3\n",
            "675: frmd3_mut has unique values: 5\n",
            "676: smad2_mut has unique values: 1\n",
            "677: sgcd_mut has unique values: 2\n",
            "678: spaca1_mut has unique values: 2\n",
            "679: rasgef1b_mut has unique values: 1\n",
            "680: hist1h2bc_mut has unique values: 3\n",
            "681: nr2f1_mut has unique values: 2\n",
            "682: klrg1_mut has unique values: 2\n",
            "683: mbl2_mut has unique values: 1\n",
            "684: mtap_mut has unique values: 2\n",
            "685: ppp2cb_mut has unique values: 2\n",
            "686: smarcd1_mut has unique values: 3\n",
            "687: nras_mut has unique values: 1\n",
            "688: ndfip1_mut has unique values: 2\n",
            "689: hras_mut has unique values: 1\n",
            "690: prps2_mut has unique values: 2\n",
            "691: smarcb1_mut has unique values: 1\n",
            "692: stmn2_mut has unique values: 1\n",
            "693: siah1_mut has unique values: 1\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data = data.set_index('patient_id')\n",
        "df_expression = data.iloc[:,\n",
        "30:519].join(data['overall_survival'], how='inner')\n",
        "df_expression"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 474
        },
        "id": "1hjadbS9q1UZ",
        "outputId": "6558f59d-aa1a-4228-c048-3a76e67e4c9c"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "             brca1   brca2   palb2    pten    tp53     atm    cdh1   chek2  \\\n",
              "patient_id                                                                   \n",
              "0          -1.3990 -0.5738 -1.6217  1.4524  0.3504  1.1517  0.0348  0.1266   \n",
              "2          -1.3800  0.2777 -1.2154  0.5296 -0.0136 -0.2659  1.3594  0.7961   \n",
              "5           0.0670 -0.8426  0.2114 -0.3326  0.5141 -0.0803  1.1398  0.4187   \n",
              "6           0.6744 -0.5428 -1.6592  0.6369  1.6708 -0.8880  1.2491 -1.1889   \n",
              "8           1.2932 -0.9039 -0.7219  0.2168  0.3484  0.3897  0.9131  0.9356   \n",
              "...            ...     ...     ...     ...     ...     ...     ...     ...   \n",
              "3436       -0.2133  0.3584 -0.6478  0.1600  0.3098 -1.2561  1.9327 -0.8945   \n",
              "3437        0.4405 -0.3679 -1.1658 -0.5324 -0.1581  0.8400  1.0522 -0.1539   \n",
              "3439       -0.5298  0.7814 -0.0255 -0.6881  0.3107  0.4589  0.3527  0.2621   \n",
              "3450        0.4089 -2.0066  0.7001  0.6006  2.0187 -1.2255  1.2207  0.8971   \n",
              "3452       -0.6113 -0.4112 -0.5326  1.2881 -1.3982 -0.2587  0.3170 -0.1585   \n",
              "\n",
              "               nbn     nf1  ...  srd5a2  srd5a3     st7    star    tnk2  \\\n",
              "patient_id                  ...                                           \n",
              "0          -0.8361 -0.8578  ... -0.0194 -1.6345 -0.2142 -0.5698 -1.1741   \n",
              "2           0.5419 -2.6059  ...  0.4534  0.4068  0.7634  0.0231  0.9121   \n",
              "5          -0.4030 -1.1305  ...  0.0668  0.8344  1.7227  0.4024 -3.7172   \n",
              "6          -0.4174 -0.6165  ... -0.7078  0.8228  0.6819 -0.1948 -2.3286   \n",
              "8           0.7675 -0.2940  ... -0.3544 -1.0150  2.2961  0.1817 -0.1572   \n",
              "...            ...     ...  ...     ...     ...     ...     ...     ...   \n",
              "3436       -0.4816 -0.6254  ... -0.5488  0.5662  0.9458  0.6918 -1.1666   \n",
              "3437       -0.9428  0.3777  ...  0.8002  0.8992  2.0359  0.2968  0.3446   \n",
              "3439       -0.7245 -0.0428  ...  0.7607  1.0186  1.0176  0.7558 -0.9423   \n",
              "3450        0.2744 -0.4220  ...  0.8414  0.3233  0.1921  2.2530 -0.0245   \n",
              "3452       -0.9302 -0.2004  ...     NaN     NaN     NaN     NaN     NaN   \n",
              "\n",
              "             tulp4  ugt2b15  ugt2b17  ugt2b7  overall_survival  \n",
              "patient_id                                                      \n",
              "0          -1.4779  -0.5954  -0.8847 -0.3354                 1  \n",
              "2          -0.9538  -0.2264   0.5398 -0.8920                 1  \n",
              "5          -1.5538   1.3701  -0.1078  0.3655                 0  \n",
              "6          -0.9924  -0.3154   0.2320 -0.4828                 1  \n",
              "8           0.0427   5.0048   3.8476  1.3223                 0  \n",
              "...            ...      ...      ...     ...               ...  \n",
              "3436        0.3349   1.2187   1.6076  0.4282                 0  \n",
              "3437        3.4423  -0.8026   0.0751 -0.9454                 1  \n",
              "3439        0.1702   1.0170   1.0133  0.6045                 1  \n",
              "3450       -0.4587   0.4001   0.0177  3.0080                 1  \n",
              "3452           NaN      NaN      NaN     NaN                 0  \n",
              "\n",
              "[715 rows x 490 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-e1c8c47f-765d-4756-b665-ab730eaee6b1\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>brca1</th>\n",
              "      <th>brca2</th>\n",
              "      <th>palb2</th>\n",
              "      <th>pten</th>\n",
              "      <th>tp53</th>\n",
              "      <th>atm</th>\n",
              "      <th>cdh1</th>\n",
              "      <th>chek2</th>\n",
              "      <th>nbn</th>\n",
              "      <th>nf1</th>\n",
              "      <th>...</th>\n",
              "      <th>srd5a2</th>\n",
              "      <th>srd5a3</th>\n",
              "      <th>st7</th>\n",
              "      <th>star</th>\n",
              "      <th>tnk2</th>\n",
              "      <th>tulp4</th>\n",
              "      <th>ugt2b15</th>\n",
              "      <th>ugt2b17</th>\n",
              "      <th>ugt2b7</th>\n",
              "      <th>overall_survival</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>patient_id</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>-1.3990</td>\n",
              "      <td>-0.5738</td>\n",
              "      <td>-1.6217</td>\n",
              "      <td>1.4524</td>\n",
              "      <td>0.3504</td>\n",
              "      <td>1.1517</td>\n",
              "      <td>0.0348</td>\n",
              "      <td>0.1266</td>\n",
              "      <td>-0.8361</td>\n",
              "      <td>-0.8578</td>\n",
              "      <td>...</td>\n",
              "      <td>-0.0194</td>\n",
              "      <td>-1.6345</td>\n",
              "      <td>-0.2142</td>\n",
              "      <td>-0.5698</td>\n",
              "      <td>-1.1741</td>\n",
              "      <td>-1.4779</td>\n",
              "      <td>-0.5954</td>\n",
              "      <td>-0.8847</td>\n",
              "      <td>-0.3354</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>-1.3800</td>\n",
              "      <td>0.2777</td>\n",
              "      <td>-1.2154</td>\n",
              "      <td>0.5296</td>\n",
              "      <td>-0.0136</td>\n",
              "      <td>-0.2659</td>\n",
              "      <td>1.3594</td>\n",
              "      <td>0.7961</td>\n",
              "      <td>0.5419</td>\n",
              "      <td>-2.6059</td>\n",
              "      <td>...</td>\n",
              "      <td>0.4534</td>\n",
              "      <td>0.4068</td>\n",
              "      <td>0.7634</td>\n",
              "      <td>0.0231</td>\n",
              "      <td>0.9121</td>\n",
              "      <td>-0.9538</td>\n",
              "      <td>-0.2264</td>\n",
              "      <td>0.5398</td>\n",
              "      <td>-0.8920</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>0.0670</td>\n",
              "      <td>-0.8426</td>\n",
              "      <td>0.2114</td>\n",
              "      <td>-0.3326</td>\n",
              "      <td>0.5141</td>\n",
              "      <td>-0.0803</td>\n",
              "      <td>1.1398</td>\n",
              "      <td>0.4187</td>\n",
              "      <td>-0.4030</td>\n",
              "      <td>-1.1305</td>\n",
              "      <td>...</td>\n",
              "      <td>0.0668</td>\n",
              "      <td>0.8344</td>\n",
              "      <td>1.7227</td>\n",
              "      <td>0.4024</td>\n",
              "      <td>-3.7172</td>\n",
              "      <td>-1.5538</td>\n",
              "      <td>1.3701</td>\n",
              "      <td>-0.1078</td>\n",
              "      <td>0.3655</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>0.6744</td>\n",
              "      <td>-0.5428</td>\n",
              "      <td>-1.6592</td>\n",
              "      <td>0.6369</td>\n",
              "      <td>1.6708</td>\n",
              "      <td>-0.8880</td>\n",
              "      <td>1.2491</td>\n",
              "      <td>-1.1889</td>\n",
              "      <td>-0.4174</td>\n",
              "      <td>-0.6165</td>\n",
              "      <td>...</td>\n",
              "      <td>-0.7078</td>\n",
              "      <td>0.8228</td>\n",
              "      <td>0.6819</td>\n",
              "      <td>-0.1948</td>\n",
              "      <td>-2.3286</td>\n",
              "      <td>-0.9924</td>\n",
              "      <td>-0.3154</td>\n",
              "      <td>0.2320</td>\n",
              "      <td>-0.4828</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>8</th>\n",
              "      <td>1.2932</td>\n",
              "      <td>-0.9039</td>\n",
              "      <td>-0.7219</td>\n",
              "      <td>0.2168</td>\n",
              "      <td>0.3484</td>\n",
              "      <td>0.3897</td>\n",
              "      <td>0.9131</td>\n",
              "      <td>0.9356</td>\n",
              "      <td>0.7675</td>\n",
              "      <td>-0.2940</td>\n",
              "      <td>...</td>\n",
              "      <td>-0.3544</td>\n",
              "      <td>-1.0150</td>\n",
              "      <td>2.2961</td>\n",
              "      <td>0.1817</td>\n",
              "      <td>-0.1572</td>\n",
              "      <td>0.0427</td>\n",
              "      <td>5.0048</td>\n",
              "      <td>3.8476</td>\n",
              "      <td>1.3223</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3436</th>\n",
              "      <td>-0.2133</td>\n",
              "      <td>0.3584</td>\n",
              "      <td>-0.6478</td>\n",
              "      <td>0.1600</td>\n",
              "      <td>0.3098</td>\n",
              "      <td>-1.2561</td>\n",
              "      <td>1.9327</td>\n",
              "      <td>-0.8945</td>\n",
              "      <td>-0.4816</td>\n",
              "      <td>-0.6254</td>\n",
              "      <td>...</td>\n",
              "      <td>-0.5488</td>\n",
              "      <td>0.5662</td>\n",
              "      <td>0.9458</td>\n",
              "      <td>0.6918</td>\n",
              "      <td>-1.1666</td>\n",
              "      <td>0.3349</td>\n",
              "      <td>1.2187</td>\n",
              "      <td>1.6076</td>\n",
              "      <td>0.4282</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3437</th>\n",
              "      <td>0.4405</td>\n",
              "      <td>-0.3679</td>\n",
              "      <td>-1.1658</td>\n",
              "      <td>-0.5324</td>\n",
              "      <td>-0.1581</td>\n",
              "      <td>0.8400</td>\n",
              "      <td>1.0522</td>\n",
              "      <td>-0.1539</td>\n",
              "      <td>-0.9428</td>\n",
              "      <td>0.3777</td>\n",
              "      <td>...</td>\n",
              "      <td>0.8002</td>\n",
              "      <td>0.8992</td>\n",
              "      <td>2.0359</td>\n",
              "      <td>0.2968</td>\n",
              "      <td>0.3446</td>\n",
              "      <td>3.4423</td>\n",
              "      <td>-0.8026</td>\n",
              "      <td>0.0751</td>\n",
              "      <td>-0.9454</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3439</th>\n",
              "      <td>-0.5298</td>\n",
              "      <td>0.7814</td>\n",
              "      <td>-0.0255</td>\n",
              "      <td>-0.6881</td>\n",
              "      <td>0.3107</td>\n",
              "      <td>0.4589</td>\n",
              "      <td>0.3527</td>\n",
              "      <td>0.2621</td>\n",
              "      <td>-0.7245</td>\n",
              "      <td>-0.0428</td>\n",
              "      <td>...</td>\n",
              "      <td>0.7607</td>\n",
              "      <td>1.0186</td>\n",
              "      <td>1.0176</td>\n",
              "      <td>0.7558</td>\n",
              "      <td>-0.9423</td>\n",
              "      <td>0.1702</td>\n",
              "      <td>1.0170</td>\n",
              "      <td>1.0133</td>\n",
              "      <td>0.6045</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3450</th>\n",
              "      <td>0.4089</td>\n",
              "      <td>-2.0066</td>\n",
              "      <td>0.7001</td>\n",
              "      <td>0.6006</td>\n",
              "      <td>2.0187</td>\n",
              "      <td>-1.2255</td>\n",
              "      <td>1.2207</td>\n",
              "      <td>0.8971</td>\n",
              "      <td>0.2744</td>\n",
              "      <td>-0.4220</td>\n",
              "      <td>...</td>\n",
              "      <td>0.8414</td>\n",
              "      <td>0.3233</td>\n",
              "      <td>0.1921</td>\n",
              "      <td>2.2530</td>\n",
              "      <td>-0.0245</td>\n",
              "      <td>-0.4587</td>\n",
              "      <td>0.4001</td>\n",
              "      <td>0.0177</td>\n",
              "      <td>3.0080</td>\n",
              "      <td>1</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3452</th>\n",
              "      <td>-0.6113</td>\n",
              "      <td>-0.4112</td>\n",
              "      <td>-0.5326</td>\n",
              "      <td>1.2881</td>\n",
              "      <td>-1.3982</td>\n",
              "      <td>-0.2587</td>\n",
              "      <td>0.3170</td>\n",
              "      <td>-0.1585</td>\n",
              "      <td>-0.9302</td>\n",
              "      <td>-0.2004</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>715 rows × 490 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-e1c8c47f-765d-4756-b665-ab730eaee6b1')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-e1c8c47f-765d-4756-b665-ab730eaee6b1 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-e1c8c47f-765d-4756-b665-ab730eaee6b1');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-12d6b44a-3c6a-4f92-8789-358a846ef2f1\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-12d6b44a-3c6a-4f92-8789-358a846ef2f1')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-12d6b44a-3c6a-4f92-8789-358a846ef2f1 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_d836e5d5-e581-4b0d-94c0-ab292e9429c2\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df_expression')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_d836e5d5-e581-4b0d-94c0-ab292e9429c2 button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df_expression');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df_expression"
            }
          },
          "metadata": {},
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Step 4: Preprocess the data (Convert categorical features to numerical using Label Encoding)\n",
        "# Create a LabelEncoder object\n",
        "label_encoder = LabelEncoder()"
      ],
      "metadata": {
        "id": "1GwTi6d-00Md"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Apply Label Encoding to each column\n",
        "for col in data.columns:\n",
        "    data[col] = label_encoder.fit_transform(data[col])"
      ],
      "metadata": {
        "id": "MCXlwVDg01Dq"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Step 5: Define Features (X) and Target (y)\n",
        "X = data.drop(columns=['type_of_breast_surgery'])  # Features\n",
        "y = data['type_of_breast_surgery']"
      ],
      "metadata": {
        "id": "n6B-aXeR01AR"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Step 6: Split the data into training and testing sets (80% train, 20% test)\n",
        "X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)"
      ],
      "metadata": {
        "id": "ShuXjmhk009A"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Step 7: Train the Random Forest Classifier\n",
        "rfc = RandomForestClassifier(n_estimators=100, random_state=42)\n",
        "rfc.fit(X_train, y_train)"
      ],
      "metadata": {
        "id": "nm7GXb5r006h",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 80
        },
        "outputId": "500c04a0-d056-4e43-9933-e133d8c9afb1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "RandomForestClassifier(random_state=42)"
            ],
            "text/html": [
              "<style>#sk-container-id-1 {\n",
              "  /* Definition of color scheme common for light and dark mode */\n",
              "  --sklearn-color-text: black;\n",
              "  --sklearn-color-line: gray;\n",
              "  /* Definition of color scheme for unfitted estimators */\n",
              "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
              "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
              "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
              "  --sklearn-color-unfitted-level-3: chocolate;\n",
              "  /* Definition of color scheme for fitted estimators */\n",
              "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
              "  --sklearn-color-fitted-level-1: #d4ebff;\n",
              "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
              "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
              "\n",
              "  /* Specific color for light theme */\n",
              "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
              "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
              "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
              "  --sklearn-color-icon: #696969;\n",
              "\n",
              "  @media (prefers-color-scheme: dark) {\n",
              "    /* Redefinition of color scheme for dark theme */\n",
              "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
              "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
              "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
              "    --sklearn-color-icon: #878787;\n",
              "  }\n",
              "}\n",
              "\n",
              "#sk-container-id-1 {\n",
              "  color: var(--sklearn-color-text);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 pre {\n",
              "  padding: 0;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 input.sk-hidden--visually {\n",
              "  border: 0;\n",
              "  clip: rect(1px 1px 1px 1px);\n",
              "  clip: rect(1px, 1px, 1px, 1px);\n",
              "  height: 1px;\n",
              "  margin: -1px;\n",
              "  overflow: hidden;\n",
              "  padding: 0;\n",
              "  position: absolute;\n",
              "  width: 1px;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-dashed-wrapped {\n",
              "  border: 1px dashed var(--sklearn-color-line);\n",
              "  margin: 0 0.4em 0.5em 0.4em;\n",
              "  box-sizing: border-box;\n",
              "  padding-bottom: 0.4em;\n",
              "  background-color: var(--sklearn-color-background);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-container {\n",
              "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
              "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
              "     so we also need the `!important` here to be able to override the\n",
              "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
              "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
              "  display: inline-block !important;\n",
              "  position: relative;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-text-repr-fallback {\n",
              "  display: none;\n",
              "}\n",
              "\n",
              "div.sk-parallel-item,\n",
              "div.sk-serial,\n",
              "div.sk-item {\n",
              "  /* draw centered vertical line to link estimators */\n",
              "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
              "  background-size: 2px 100%;\n",
              "  background-repeat: no-repeat;\n",
              "  background-position: center center;\n",
              "}\n",
              "\n",
              "/* Parallel-specific style estimator block */\n",
              "\n",
              "#sk-container-id-1 div.sk-parallel-item::after {\n",
              "  content: \"\";\n",
              "  width: 100%;\n",
              "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
              "  flex-grow: 1;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-parallel {\n",
              "  display: flex;\n",
              "  align-items: stretch;\n",
              "  justify-content: center;\n",
              "  background-color: var(--sklearn-color-background);\n",
              "  position: relative;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-parallel-item {\n",
              "  display: flex;\n",
              "  flex-direction: column;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
              "  align-self: flex-end;\n",
              "  width: 50%;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
              "  align-self: flex-start;\n",
              "  width: 50%;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
              "  width: 0;\n",
              "}\n",
              "\n",
              "/* Serial-specific style estimator block */\n",
              "\n",
              "#sk-container-id-1 div.sk-serial {\n",
              "  display: flex;\n",
              "  flex-direction: column;\n",
              "  align-items: center;\n",
              "  background-color: var(--sklearn-color-background);\n",
              "  padding-right: 1em;\n",
              "  padding-left: 1em;\n",
              "}\n",
              "\n",
              "\n",
              "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
              "clickable and can be expanded/collapsed.\n",
              "- Pipeline and ColumnTransformer use this feature and define the default style\n",
              "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
              "*/\n",
              "\n",
              "/* Pipeline and ColumnTransformer style (default) */\n",
              "\n",
              "#sk-container-id-1 div.sk-toggleable {\n",
              "  /* Default theme specific background. It is overwritten whether we have a\n",
              "  specific estimator or a Pipeline/ColumnTransformer */\n",
              "  background-color: var(--sklearn-color-background);\n",
              "}\n",
              "\n",
              "/* Toggleable label */\n",
              "#sk-container-id-1 label.sk-toggleable__label {\n",
              "  cursor: pointer;\n",
              "  display: block;\n",
              "  width: 100%;\n",
              "  margin-bottom: 0;\n",
              "  padding: 0.5em;\n",
              "  box-sizing: border-box;\n",
              "  text-align: center;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
              "  /* Arrow on the left of the label */\n",
              "  content: \"▸\";\n",
              "  float: left;\n",
              "  margin-right: 0.25em;\n",
              "  color: var(--sklearn-color-icon);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
              "  color: var(--sklearn-color-text);\n",
              "}\n",
              "\n",
              "/* Toggleable content - dropdown */\n",
              "\n",
              "#sk-container-id-1 div.sk-toggleable__content {\n",
              "  max-height: 0;\n",
              "  max-width: 0;\n",
              "  overflow: hidden;\n",
              "  text-align: left;\n",
              "  /* unfitted */\n",
              "  background-color: var(--sklearn-color-unfitted-level-0);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
              "  /* fitted */\n",
              "  background-color: var(--sklearn-color-fitted-level-0);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-toggleable__content pre {\n",
              "  margin: 0.2em;\n",
              "  border-radius: 0.25em;\n",
              "  color: var(--sklearn-color-text);\n",
              "  /* unfitted */\n",
              "  background-color: var(--sklearn-color-unfitted-level-0);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
              "  /* unfitted */\n",
              "  background-color: var(--sklearn-color-fitted-level-0);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
              "  /* Expand drop-down */\n",
              "  max-height: 200px;\n",
              "  max-width: 100%;\n",
              "  overflow: auto;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
              "  content: \"▾\";\n",
              "}\n",
              "\n",
              "/* Pipeline/ColumnTransformer-specific style */\n",
              "\n",
              "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
              "  color: var(--sklearn-color-text);\n",
              "  background-color: var(--sklearn-color-unfitted-level-2);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
              "  background-color: var(--sklearn-color-fitted-level-2);\n",
              "}\n",
              "\n",
              "/* Estimator-specific style */\n",
              "\n",
              "/* Colorize estimator box */\n",
              "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
              "  /* unfitted */\n",
              "  background-color: var(--sklearn-color-unfitted-level-2);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
              "  /* fitted */\n",
              "  background-color: var(--sklearn-color-fitted-level-2);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
              "#sk-container-id-1 div.sk-label label {\n",
              "  /* The background is the default theme color */\n",
              "  color: var(--sklearn-color-text-on-default-background);\n",
              "}\n",
              "\n",
              "/* On hover, darken the color of the background */\n",
              "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
              "  color: var(--sklearn-color-text);\n",
              "  background-color: var(--sklearn-color-unfitted-level-2);\n",
              "}\n",
              "\n",
              "/* Label box, darken color on hover, fitted */\n",
              "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
              "  color: var(--sklearn-color-text);\n",
              "  background-color: var(--sklearn-color-fitted-level-2);\n",
              "}\n",
              "\n",
              "/* Estimator label */\n",
              "\n",
              "#sk-container-id-1 div.sk-label label {\n",
              "  font-family: monospace;\n",
              "  font-weight: bold;\n",
              "  display: inline-block;\n",
              "  line-height: 1.2em;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-label-container {\n",
              "  text-align: center;\n",
              "}\n",
              "\n",
              "/* Estimator-specific */\n",
              "#sk-container-id-1 div.sk-estimator {\n",
              "  font-family: monospace;\n",
              "  border: 1px dotted var(--sklearn-color-border-box);\n",
              "  border-radius: 0.25em;\n",
              "  box-sizing: border-box;\n",
              "  margin-bottom: 0.5em;\n",
              "  /* unfitted */\n",
              "  background-color: var(--sklearn-color-unfitted-level-0);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-estimator.fitted {\n",
              "  /* fitted */\n",
              "  background-color: var(--sklearn-color-fitted-level-0);\n",
              "}\n",
              "\n",
              "/* on hover */\n",
              "#sk-container-id-1 div.sk-estimator:hover {\n",
              "  /* unfitted */\n",
              "  background-color: var(--sklearn-color-unfitted-level-2);\n",
              "}\n",
              "\n",
              "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
              "  /* fitted */\n",
              "  background-color: var(--sklearn-color-fitted-level-2);\n",
              "}\n",
              "\n",
              "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
              "\n",
              "/* Common style for \"i\" and \"?\" */\n",
              "\n",
              ".sk-estimator-doc-link,\n",
              "a:link.sk-estimator-doc-link,\n",
              "a:visited.sk-estimator-doc-link {\n",
              "  float: right;\n",
              "  font-size: smaller;\n",
              "  line-height: 1em;\n",
              "  font-family: monospace;\n",
              "  background-color: var(--sklearn-color-background);\n",
              "  border-radius: 1em;\n",
              "  height: 1em;\n",
              "  width: 1em;\n",
              "  text-decoration: none !important;\n",
              "  margin-left: 1ex;\n",
              "  /* unfitted */\n",
              "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
              "  color: var(--sklearn-color-unfitted-level-1);\n",
              "}\n",
              "\n",
              ".sk-estimator-doc-link.fitted,\n",
              "a:link.sk-estimator-doc-link.fitted,\n",
              "a:visited.sk-estimator-doc-link.fitted {\n",
              "  /* fitted */\n",
              "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
              "  color: var(--sklearn-color-fitted-level-1);\n",
              "}\n",
              "\n",
              "/* On hover */\n",
              "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
              ".sk-estimator-doc-link:hover,\n",
              "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
              ".sk-estimator-doc-link:hover {\n",
              "  /* unfitted */\n",
              "  background-color: var(--sklearn-color-unfitted-level-3);\n",
              "  color: var(--sklearn-color-background);\n",
              "  text-decoration: none;\n",
              "}\n",
              "\n",
              "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
              ".sk-estimator-doc-link.fitted:hover,\n",
              "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
              ".sk-estimator-doc-link.fitted:hover {\n",
              "  /* fitted */\n",
              "  background-color: var(--sklearn-color-fitted-level-3);\n",
              "  color: var(--sklearn-color-background);\n",
              "  text-decoration: none;\n",
              "}\n",
              "\n",
              "/* Span, style for the box shown on hovering the info icon */\n",
              ".sk-estimator-doc-link span {\n",
              "  display: none;\n",
              "  z-index: 9999;\n",
              "  position: relative;\n",
              "  font-weight: normal;\n",
              "  right: .2ex;\n",
              "  padding: .5ex;\n",
              "  margin: .5ex;\n",
              "  width: min-content;\n",
              "  min-width: 20ex;\n",
              "  max-width: 50ex;\n",
              "  color: var(--sklearn-color-text);\n",
              "  box-shadow: 2pt 2pt 4pt #999;\n",
              "  /* unfitted */\n",
              "  background: var(--sklearn-color-unfitted-level-0);\n",
              "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
              "}\n",
              "\n",
              ".sk-estimator-doc-link.fitted span {\n",
              "  /* fitted */\n",
              "  background: var(--sklearn-color-fitted-level-0);\n",
              "  border: var(--sklearn-color-fitted-level-3);\n",
              "}\n",
              "\n",
              ".sk-estimator-doc-link:hover span {\n",
              "  display: block;\n",
              "}\n",
              "\n",
              "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
              "\n",
              "#sk-container-id-1 a.estimator_doc_link {\n",
              "  float: right;\n",
              "  font-size: 1rem;\n",
              "  line-height: 1em;\n",
              "  font-family: monospace;\n",
              "  background-color: var(--sklearn-color-background);\n",
              "  border-radius: 1rem;\n",
              "  height: 1rem;\n",
              "  width: 1rem;\n",
              "  text-decoration: none;\n",
              "  /* unfitted */\n",
              "  color: var(--sklearn-color-unfitted-level-1);\n",
              "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
              "  /* fitted */\n",
              "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
              "  color: var(--sklearn-color-fitted-level-1);\n",
              "}\n",
              "\n",
              "/* On hover */\n",
              "#sk-container-id-1 a.estimator_doc_link:hover {\n",
              "  /* unfitted */\n",
              "  background-color: var(--sklearn-color-unfitted-level-3);\n",
              "  color: var(--sklearn-color-background);\n",
              "  text-decoration: none;\n",
              "}\n",
              "\n",
              "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
              "  /* fitted */\n",
              "  background-color: var(--sklearn-color-fitted-level-3);\n",
              "}\n",
              "</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>RandomForestClassifier(random_state=42)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow fitted\">&nbsp;&nbsp;RandomForestClassifier<a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://scikit-learn.org/1.5/modules/generated/sklearn.ensemble.RandomForestClassifier.html\">?<span>Documentation for RandomForestClassifier</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></label><div class=\"sk-toggleable__content fitted\"><pre>RandomForestClassifier(random_state=42)</pre></div> </div></div></div></div>"
            ]
          },
          "metadata": {},
          "execution_count": 14
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Step 8: Make predictions on the test set\n",
        "y_pred = rfc.predict(X_test)"
      ],
      "metadata": {
        "id": "UAxSfDrW003m"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Classification report\n",
        "print(\"\\nClassification Report:\")\n",
        "print(classification_report(y_test, y_pred))"
      ],
      "metadata": {
        "id": "BI2cYREP1JiO",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "7d4f37de-d077-4585-9de3-e907256a3aa1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Classification Report:\n",
            "              precision    recall  f1-score   support\n",
            "\n",
            "           0       0.82      0.23      0.36        78\n",
            "           1       0.63      0.96      0.76       110\n",
            "           2       0.00      0.00      0.00         3\n",
            "\n",
            "    accuracy                           0.65       191\n",
            "   macro avg       0.48      0.40      0.37       191\n",
            "weighted avg       0.70      0.65      0.58       191\n",
            "\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
            "/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n",
            "/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1531: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.\n",
            "  _warn_prf(average, modifier, f\"{metric.capitalize()} is\", len(result))\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Step 9: Evaluate the model\n",
        "accuracy = accuracy_score(y_test, y_pred)\n",
        "precision = precision_score(y_test, y_pred)\n",
        "recall = recall_score(y_test, y_pred)\n",
        "error_rate = 1 - accuracy\n",
        "print(f\"Model Accuracy: {accuracy * 100:.2f}%\")\n",
        "print(f\"Precision: {precision * 100:.2f}%\")\n",
        "print(f\"Recall: {recall * 100:.2f}%\")\n",
        "print(f\"Error Rate: {error_rate * 100:.2f}%\")"
      ],
      "metadata": {
        "id": "D4YnPVz51GRP",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 391
        },
        "collapsed": true,
        "outputId": "ad2b9d7f-d842-47fb-c80c-98cbd704a1ba"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "ValueError",
          "evalue": "Target is multiclass but average='binary'. Please choose another average setting, one of [None, 'micro', 'macro', 'weighted'].",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-19-b09f7894f146>\u001b[0m in \u001b[0;36m<cell line: 3>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m# Step 9: Evaluate the model\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0maccuracy\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0maccuracy_score\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mprecision\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mprecision_score\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0mrecall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mrecall_score\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_test\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0merror_rate\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m1\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0maccuracy\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_param_validation.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    211\u001b[0m                     )\n\u001b[1;32m    212\u001b[0m                 ):\n\u001b[0;32m--> 213\u001b[0;31m                     \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    214\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mInvalidParameterError\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    215\u001b[0m                 \u001b[0;31m# When the function is just a wrapper around an estimator, we allow\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py\u001b[0m in \u001b[0;36mprecision_score\u001b[0;34m(y_true, y_pred, labels, pos_label, average, sample_weight, zero_division)\u001b[0m\n\u001b[1;32m   2202\u001b[0m     \u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0.5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1.\u001b[0m \u001b[0;34m,\u001b[0m \u001b[0;36m1.\u001b[0m \u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2203\u001b[0m     \"\"\"\n\u001b[0;32m-> 2204\u001b[0;31m     p, _, _, _ = precision_recall_fscore_support(\n\u001b[0m\u001b[1;32m   2205\u001b[0m         \u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2206\u001b[0m         \u001b[0my_pred\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/utils/_param_validation.py\u001b[0m in \u001b[0;36mwrapper\u001b[0;34m(*args, **kwargs)\u001b[0m\n\u001b[1;32m    184\u001b[0m             \u001b[0mglobal_skip_validation\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_config\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"skip_parameter_validation\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    185\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mglobal_skip_validation\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 186\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mfunc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    187\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    188\u001b[0m             \u001b[0mfunc_sig\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msignature\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfunc\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py\u001b[0m in \u001b[0;36mprecision_recall_fscore_support\u001b[0;34m(y_true, y_pred, beta, labels, pos_label, average, warn_for, sample_weight, zero_division)\u001b[0m\n\u001b[1;32m   1787\u001b[0m     \"\"\"\n\u001b[1;32m   1788\u001b[0m     \u001b[0m_check_zero_division\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mzero_division\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1789\u001b[0;31m     \u001b[0mlabels\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_check_set_wise_labels\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0my_true\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0my_pred\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maverage\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlabels\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpos_label\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1790\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1791\u001b[0m     \u001b[0;31m# Calculate tp_sum, pred_sum, true_sum ###\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py\u001b[0m in \u001b[0;36m_check_set_wise_labels\u001b[0;34m(y_true, y_pred, average, labels, pos_label)\u001b[0m\n\u001b[1;32m   1576\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0my_type\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"multiclass\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1577\u001b[0m                 \u001b[0maverage_options\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mremove\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"samples\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1578\u001b[0;31m             raise ValueError(\n\u001b[0m\u001b[1;32m   1579\u001b[0m                 \u001b[0;34m\"Target is %s but average='binary'. Please \"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1580\u001b[0m                 \u001b[0;34m\"choose another average setting, one of %r.\"\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0my_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maverage_options\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mValueError\u001b[0m: Target is multiclass but average='binary'. Please choose another average setting, one of [None, 'micro', 'macro', 'weighted']."
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Confusion matrix\n",
        "conf_matrix = confusion_matrix(y_test, y_pred)\n",
        "print(\"\\nConfusion Matrix:\")\n",
        "print(conf_matrix)"
      ],
      "metadata": {
        "id": "D9UrBYtT1GNr",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "66fe9983-f92a-439d-83c0-885e9657dd15"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\n",
            "Confusion Matrix:\n",
            "[[ 18  60   0]\n",
            " [  4 106   0]\n",
            " [  0   3   0]]\n"
          ]
        }
      ]
    }
  ]
}